// @hubspot/cli/commands/cms/delete.js
import { deleteFile } from '@hubspot/local-dev-lib/api/fileMapper';
import { logError, ApiErrorContext } from '../../lib/errorHandlers/index.js';
import { trackCommandUsage } from '../../lib/usageTracking.js';
import { commands } from '../../lang/en.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
import { uiLogger } from '../../lib/ui/logger.js';
const command = 'delete <path>';
const describe = commands.cms.subcommands.delete.describe;
async function handler(args) {
    const { path: hsPath, derivedAccountId } = args;
    trackCommandUsage('delete', undefined, derivedAccountId);
    try {
        await deleteFile(derivedAccountId, hsPath);
        uiLogger.log(commands.cms.subcommands.delete.deleted(hsPath, derivedAccountId));
    }
    catch (error) {
        uiLogger.error(commands.cms.subcommands.delete.errors.deleteFailed(hsPath, derivedAccountId));
        logError(error, new ApiErrorContext({
            accountId: derivedAccountId,
            request: hsPath,
        }));
    }
}
function cmsDeleteBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.delete.positionals.path,
        type: 'string',
    });
    return yargs;
}
const builder = makeYargsBuilder(cmsDeleteBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const cmsDeleteCommand = {
    command,
    describe,
    handler,
    builder,
};
export default cmsDeleteCommand;
