// @hubspot/cli/commands/cms/lighthouseScore.js
import SpinniesManager from '../../lib/ui/SpinniesManager.js';
import { getTableContents, getTableHeader } from '../../lib/ui/table.js';
import { promptUser } from '../../lib/prompts/promptUtils.js';
import { commands } from '../../lang/en.js';
import { uiLogger } from '../../lib/ui/logger.js';
import { fetchThemes } from '@hubspot/local-dev-lib/api/designManager';
import { requestLighthouseScore, getLighthouseScoreStatus, getLighthouseScore, } from '@hubspot/local-dev-lib/api/lighthouseScore';
import { HUBSPOT_FOLDER, MARKETPLACE_FOLDER } from '../../lib/constants.js';
import { uiLink } from '../../lib/ui/index.js';
import { EXIT_CODES } from '../../lib/enums/exitCodes.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
const DEFAULT_TABLE_HEADER = [
    'Accessibility',
    'Best practices',
    'Performace',
    'PWA',
    'SEO',
];
const EMPTY_SCORE = {
    accessibilityScore: 0,
    bestPracticesScore: 0,
    performanceScore: 0,
    pwaScore: 0,
    seoScore: 0,
    runWarnings: [],
    auditDetails: null,
    emulatedFormFactor: '',
    templatePath: null,
    link: null,
};
const command = 'lighthouse-score [--theme]';
const describe = undefined;
async function selectTheme(accountId) {
    let themes = [];
    try {
        const { data: result } = await fetchThemes(accountId, {
            limit: 500,
            sorting: 'MOST_USED',
        });
        if (result && result.objects) {
            themes = result.objects
                .map(({ theme }) => theme.path)
                .filter(themePath => !themePath.startsWith(HUBSPOT_FOLDER) &&
                !themePath.startsWith(MARKETPLACE_FOLDER));
        }
    }
    catch (err) {
        uiLogger.error(commands.cms.subcommands.lighthouseScore.errors.failedToFetchThemes);
        process.exit(EXIT_CODES.ERROR);
    }
    const { theme: selectedTheme } = await promptUser([
        {
            type: 'list',
            name: 'theme',
            message: commands.cms.subcommands.lighthouseScore.info.promptMessage,
            choices: themes,
        },
    ]);
    return selectedTheme;
}
async function handler(args) {
    const { target, verbose, theme, derivedAccountId } = args;
    const includeDesktopScore = target === 'desktop' || !verbose;
    const includeMobileScore = target === 'mobile' || !verbose;
    let themeToCheck = theme;
    if (themeToCheck) {
        let isValidTheme = true;
        try {
            const { data: result } = await fetchThemes(derivedAccountId, {
                name: encodeURIComponent(themeToCheck),
            });
            isValidTheme = result && !!result.total;
        }
        catch (err) {
            isValidTheme = false;
        }
        if (!isValidTheme) {
            uiLogger.error(commands.cms.subcommands.lighthouseScore.errors.themeNotFound(themeToCheck));
            process.exit(EXIT_CODES.ERROR);
        }
    }
    else {
        themeToCheck = await selectTheme(derivedAccountId);
        uiLogger.log('');
    }
    // Kick off the scoring
    let requestResult;
    try {
        const { data } = await requestLighthouseScore(derivedAccountId, {
            themePath: themeToCheck,
        });
        requestResult = data;
    }
    catch (err) {
        uiLogger.debug(err);
    }
    if (!requestResult || !requestResult.mobileId || !requestResult.desktopId) {
        uiLogger.error(commands.cms.subcommands.lighthouseScore.errors.failedToGetLighthouseScore);
        process.exit(EXIT_CODES.ERROR);
    }
    // Poll till scoring is finished
    try {
        SpinniesManager.init();
        SpinniesManager.add('lighthouseScore', {
            text: commands.cms.subcommands.lighthouseScore.info.generatingScore(themeToCheck),
        });
        async function checkScoreStatus() {
            let desktopScoreStatus = 'COMPLETED';
            if (includeDesktopScore) {
                const { data } = await getLighthouseScoreStatus(derivedAccountId, {
                    themeId: requestResult.desktopId,
                });
                desktopScoreStatus = data;
            }
            let mobileScoreStatus = 'COMPLETED';
            if (includeDesktopScore) {
                const { data } = await getLighthouseScoreStatus(derivedAccountId, {
                    themeId: requestResult.mobileId,
                });
                mobileScoreStatus = data;
            }
            if (desktopScoreStatus === 'REQUESTED' ||
                mobileScoreStatus === 'REQUESTED') {
                await new Promise(resolve => setTimeout(resolve, 2000));
                await checkScoreStatus();
            }
        }
        await checkScoreStatus();
        SpinniesManager.remove('lighthouseScore');
    }
    catch (err) {
        uiLogger.debug(err);
        process.exit(EXIT_CODES.ERROR);
    }
    // Fetch the scoring results
    let desktopScoreResult;
    let mobileScoreResult;
    let verboseViewAverageScoreResult;
    try {
        const params = { isAverage: !verbose };
        if (includeDesktopScore) {
            const { data } = await getLighthouseScore(derivedAccountId, {
                ...params,
                desktopId: requestResult.desktopId,
            });
            desktopScoreResult = data;
        }
        if (includeMobileScore) {
            const { data } = await getLighthouseScore(derivedAccountId, {
                ...params,
                mobileId: requestResult.mobileId,
            });
            mobileScoreResult = data;
        }
        // This is needed to show the average scores above the verbose output
        if (verbose) {
            const { data } = await getLighthouseScore(derivedAccountId, {
                ...params,
                isAverage: true,
                desktopId: includeDesktopScore ? requestResult.desktopId : undefined,
                mobileId: includeMobileScore ? requestResult.mobileId : undefined,
            });
            verboseViewAverageScoreResult = data;
        }
    }
    catch (err) {
        uiLogger.error(commands.cms.subcommands.lighthouseScore.errors.failedToGetLighthouseScore);
        process.exit(EXIT_CODES.ERROR);
    }
    if (verbose) {
        const scoreResult = target === 'desktop' ? desktopScoreResult : mobileScoreResult;
        if (!verboseViewAverageScoreResult || !scoreResult) {
            uiLogger.error(commands.cms.subcommands.lighthouseScore.errors
                .failedToGetLighthouseScore);
            process.exit(EXIT_CODES.ERROR);
        }
        uiLogger.log(commands.cms.subcommands.lighthouseScore.info.themeToCheckTitle(themeToCheck, target));
        const tableHeader = getTableHeader(DEFAULT_TABLE_HEADER);
        const scores = verboseViewAverageScoreResult.scores
            ? verboseViewAverageScoreResult.scores[0]
            : EMPTY_SCORE;
        const averageTableData = [
            scores.accessibilityScore,
            scores.bestPracticesScore,
            scores.performanceScore,
            scores.pwaScore,
            scores.seoScore,
        ];
        uiLogger.log(getTableContents([tableHeader, averageTableData], {
            border: { bodyLeft: '  ' },
        }));
        uiLogger.log(commands.cms.subcommands.lighthouseScore.info.pageTemplateScoreTitle);
        const table2Header = getTableHeader([
            'Template path',
            ...DEFAULT_TABLE_HEADER,
        ]);
        const templateTableData = scoreResult.scores.map(score => {
            return [
                score.templatePath,
                score.accessibilityScore,
                score.bestPracticesScore,
                score.performanceScore,
                score.pwaScore,
                score.seoScore,
            ];
        });
        uiLogger.log(getTableContents([table2Header, ...templateTableData], {
            border: { bodyLeft: '  ' },
        }));
        uiLogger.log(commands.cms.subcommands.lighthouseScore.info.lighthouseLinksTitle);
        scoreResult.scores.forEach(score => {
            if (score.templatePath && score.link) {
                uiLogger.log(` ${uiLink(score.templatePath, score.link)}`);
            }
        });
        if (scoreResult.failedTemplatePaths.length) {
            uiLogger.log('');
            uiLogger.error(commands.cms.subcommands.lighthouseScore.info.failedTemplatePathsTitle);
            scoreResult.failedTemplatePaths.forEach(failedTemplatePath => {
                uiLogger.log(` ${failedTemplatePath}`);
            });
        }
        uiLogger.log('');
        uiLogger.info(commands.cms.subcommands.lighthouseScore.info.targetDeviceNote(target));
    }
    else {
        uiLogger.log(commands.cms.subcommands.lighthouseScore.info.themeTitle(themeToCheck));
        const tableHeader = getTableHeader(['Target', ...DEFAULT_TABLE_HEADER]);
        const getTableData = (target, scoreResult) => {
            const scores = scoreResult?.scores ? scoreResult.scores[0] : EMPTY_SCORE;
            return [
                target,
                scores.accessibilityScore,
                scores.bestPracticesScore,
                scores.performanceScore,
                scores.pwaScore,
                scores.seoScore,
            ];
        };
        const tableData = [
            getTableData('desktop', desktopScoreResult),
            getTableData('mobile', mobileScoreResult),
        ];
        uiLogger.log(getTableContents([tableHeader, ...tableData], {
            border: { bodyLeft: '  ' },
        }));
        uiLogger.info(commands.cms.subcommands.lighthouseScore.info.verboseOptionNote);
    }
    uiLogger.log('');
    uiLogger.log(commands.cms.subcommands.lighthouseScore.info.poweredByLink);
    process.exit(EXIT_CODES.SUCCESS);
}
function cmslighthouseScoreBuilder(yargs) {
    yargs.option('theme', {
        describe: commands.cms.subcommands.lighthouseScore.options.theme.describe,
        type: 'string',
    });
    yargs.option('target', {
        describe: commands.cms.subcommands.lighthouseScore.options.target.describe,
        type: 'string',
        choices: ['desktop', 'mobile'],
        default: 'desktop',
    });
    yargs.option('verbose', {
        describe: commands.cms.subcommands.lighthouseScore.options.verbose.describe,
        boolean: true,
        default: false,
    });
    yargs.example([
        [
            '$0 cms lighthouse-score --theme=my-theme',
            commands.cms.subcommands.lighthouseScore.examples.default,
        ],
    ]);
    return yargs;
}
const builder = makeYargsBuilder(cmslighthouseScoreBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const cmslighthouseScoreCommand = {
    command,
    describe,
    handler,
    builder,
};
export default cmslighthouseScoreCommand;
