// @hubspot/cli/commands/cms/lint.js
import { lint } from '@hubspot/local-dev-lib/cms/validate';
import { logError } from '../../lib/errorHandlers/index.js';
import { trackCommandUsage } from '../../lib/usageTracking.js';
import { commands } from '../../lang/en.js';
import { resolveLocalPath } from '../../lib/filesystem.js';
import { EXIT_CODES } from '../../lib/enums/exitCodes.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
import { uiLogger } from '../../lib/ui/logger.js';
const command = 'lint <path>';
// Hiding since this command is still experimental
const describe = undefined; //'Lint a file or folder for HubL syntax';
function getErrorsFromHublValidationObject(validation) {
    return ((validation && validation.meta && validation.meta.template_errors) || []);
}
function printHublValidationError(err) {
    const { severity, message, lineno, startPosition } = err;
    if (severity === 'FATAL') {
        uiLogger.error(`[${lineno}, ${startPosition}]: ${message}`);
    }
    else {
        uiLogger.warn(`[${lineno}, ${startPosition}]: ${message}`);
    }
}
function printHublValidationResult({ file, validation }) {
    let count = 0;
    if (!validation) {
        return count;
    }
    const errors = getErrorsFromHublValidationObject(validation);
    if (!errors.length) {
        return count;
    }
    uiLogger.group(file);
    errors.forEach(err => {
        if (err.reason !== 'SYNTAX_ERROR') {
            return;
        }
        ++count;
        printHublValidationError(err);
    });
    uiLogger.groupEnd();
    return count;
}
async function handler(args) {
    const { path: lintPath, derivedAccountId } = args;
    const localPath = resolveLocalPath(lintPath);
    const groupName = commands.cms.subcommands.lint.groupName(localPath);
    trackCommandUsage('lint', undefined, derivedAccountId);
    uiLogger.group(groupName);
    let count = 0;
    try {
        await lint(derivedAccountId, localPath, (result) => {
            return (count += printHublValidationResult(result));
        });
    }
    catch (err) {
        uiLogger.groupEnd();
        logError(err, { accountId: derivedAccountId });
        process.exit(EXIT_CODES.ERROR);
    }
    uiLogger.groupEnd();
    uiLogger.log(commands.cms.subcommands.lint.issuesFound(count));
}
function lintBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.lint.positionals.path.describe,
        required: true,
        type: 'string',
    });
    return yargs;
}
const builder = makeYargsBuilder(lintBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
});
const lintCommand = {
    command,
    describe,
    handler,
    builder,
};
export default lintCommand;
