// @hubspot/cli/commands/cms/mv.js
import { moveFile } from '@hubspot/local-dev-lib/api/fileMapper';
import { isSpecifiedError } from '@hubspot/local-dev-lib/errors/index';
import { logError, ApiErrorContext } from '../../lib/errorHandlers/index.js';
import { trackCommandUsage } from '../../lib/usageTracking.js';
import { isPathFolder } from '../../lib/filesystem.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
import { uiLogger } from '../../lib/ui/logger.js';
import { commands } from '../../lang/en.js';
function getCorrectedDestPath(srcPath, destPath) {
    if (!isPathFolder(srcPath)) {
        return destPath;
    }
    // Makes sure that nested folders are moved independently
    return `${destPath}/${srcPath.split('/').pop()}`;
}
const command = 'mv <srcPath> <destPath>';
const describe = commands.cms.subcommands.mv.describe;
async function handler(args) {
    const { srcPath, destPath, derivedAccountId } = args;
    trackCommandUsage('mv', undefined, derivedAccountId);
    try {
        await moveFile(derivedAccountId, srcPath, getCorrectedDestPath(srcPath, destPath));
        uiLogger.success(commands.cms.subcommands.mv.move(srcPath, destPath, derivedAccountId));
    }
    catch (error) {
        uiLogger.error(commands.cms.subcommands.mv.errors.moveFailed(srcPath, destPath, derivedAccountId));
        if (isSpecifiedError(error, { statusCode: 409 })) {
            uiLogger.error(commands.cms.subcommands.mv.errors.sourcePathExists(srcPath, destPath));
        }
        else {
            logError(error, new ApiErrorContext({
                accountId: derivedAccountId,
            }));
        }
    }
}
function cmsMvBuilder(yargs) {
    yargs.positional('srcPath', {
        describe: commands.cms.subcommands.mv.positionals.srcPath.describe,
        type: 'string',
    });
    yargs.positional('destPath', {
        describe: commands.cms.subcommands.mv.positionals.destPath.describe,
        type: 'string',
    });
    return yargs;
}
const builder = makeYargsBuilder(cmsMvBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const cmsMvCommand = {
    command,
    describe,
    handler,
    builder,
};
export default cmsMvCommand;
