// @hubspot/project-parsing-lib/src/lib/constants.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.allowedTopLevelFields = exports.ProfileVariableTypes = exports.allowedSubComponentDirectories = exports.allowedComponentDirectories = exports.ProjectStructure = exports.allowedCmsAssetSubComponentDirs = exports.allowedThemeSubComponentsDirs = exports.allowedAppSubComponentsDirs = exports.packageLockJson = exports.packageJson = exports.hsProjectJsonFilename = exports.profileFilePrefix = exports.metafileExtension = exports.internalTypeToUserFacing = exports.userFacingToInternalType = exports.Components = exports.AutoGeneratedComponentTypes = exports.AppFunctionsPackageKey = exports.WorkflowActionsKey = exports.WebhooksJournalKey = exports.WebhooksKey = exports.VideoConferencingKey = exports.TelemetryKey = exports.SettingsKey = exports.MediaBridgeKey = exports.MCPRegistryKey = exports.MarketingEventsKey = exports.SCIMKey = exports.CardsKey = exports.CallingKey = exports.AppObjectAssociationKey = exports.AppObjectKey = exports.PagesKey = exports.AppFunctionsKey = exports.AppEventsKey = exports.CmsAssetsKey = exports.ThemeKey = exports.AppKey = void 0;
// Top Level Component types
const path = __importStar(require("path"));
// Component types
exports.AppKey = 'app';
exports.ThemeKey = 'theme';
exports.CmsAssetsKey = 'cms-assets';
// Sub-Component types
exports.AppEventsKey = 'app-event';
exports.AppFunctionsKey = 'app-function';
exports.PagesKey = 'page';
exports.AppObjectKey = 'app-object';
exports.AppObjectAssociationKey = 'app-object-association';
exports.CallingKey = 'calling';
exports.CardsKey = 'card';
exports.SCIMKey = 'scim';
exports.MarketingEventsKey = 'marketing-event';
exports.MCPRegistryKey = 'mcp-registry';
exports.MediaBridgeKey = 'media-bridge';
exports.SettingsKey = 'settings';
exports.TelemetryKey = 'telemetry';
exports.VideoConferencingKey = 'video-conferencing';
exports.WebhooksKey = 'webhooks';
exports.WebhooksJournalKey = 'webhooks-journal';
exports.WorkflowActionsKey = 'workflow-action';
// Auto-generated component types
exports.AppFunctionsPackageKey = 'serverless-package';
exports.AutoGeneratedComponentTypes = [exports.AppFunctionsPackageKey];
const TopLevelComponentFields = {
    isToplevel: true,
    userFriendlyType: 'component',
    userFriendlyTypePlural: 'components',
};
const SubComponentFields = {
    isToplevel: false,
    userFriendlyType: 'feature',
    userFriendlyTypePlural: 'features',
};
exports.Components = {
    [exports.AppKey]: {
        dir: exports.AppKey,
        userFriendlyName: 'App',
        ...TopLevelComponentFields,
        singularComponent: true,
    },
    [exports.AppObjectKey]: {
        dir: 'app-objects',
        parentComponent: exports.AppKey,
        ...SubComponentFields,
        userFriendlyName: 'App Object',
    },
    [exports.AppObjectAssociationKey]: {
        dir: 'app-object-associations',
        parentComponent: exports.AppKey,
        ...SubComponentFields,
        userFriendlyName: 'App Object Association',
    },
    [exports.PagesKey]: {
        dir: 'pages',
        parentComponent: exports.AppKey,
        ...SubComponentFields,
        userFriendlyName: 'Page',
        singularComponent: true,
    },
    [exports.ThemeKey]: {
        dir: exports.ThemeKey,
        userFriendlyName: 'Theme',
        ...TopLevelComponentFields,
    },
    [exports.CmsAssetsKey]: {
        dir: exports.CmsAssetsKey,
        userFriendlyName: 'CMS Asset',
        ...TopLevelComponentFields,
    },
    [exports.CallingKey]: {
        dir: exports.CallingKey,
        parentComponent: exports.AppKey,
        userFriendlyName: 'Calling',
        ...SubComponentFields,
        singularComponent: true,
    },
    [exports.CardsKey]: {
        dir: 'cards',
        parentComponent: exports.AppKey,
        userFriendlyName: 'Card',
        ...SubComponentFields,
    },
    [exports.AppFunctionsKey]: {
        dir: 'functions',
        parentComponent: exports.AppKey,
        userFriendlyName: 'App Function',
        ...SubComponentFields,
    },
    [exports.SCIMKey]: {
        dir: exports.SCIMKey,
        parentComponent: exports.AppKey,
        userFriendlyName: 'SCIM',
        singularComponent: true,
        ...SubComponentFields,
    },
    [exports.SettingsKey]: {
        dir: 'settings',
        parentComponent: exports.AppKey,
        userFriendlyName: 'Settings Extension',
        ...SubComponentFields,
        singularComponent: true,
    },
    [exports.MarketingEventsKey]: {
        dir: 'marketing-events',
        parentComponent: exports.AppKey,
        userFriendlyName: 'Marketing Event',
        ...SubComponentFields,
    },
    MediaBridgeKey: {
        dir: exports.MediaBridgeKey,
        parentComponent: exports.AppKey,
        userFriendlyName: 'Media Bridge',
        ...SubComponentFields,
    },
    [exports.AppEventsKey]: {
        dir: 'app-events',
        parentComponent: exports.AppKey,
        userFriendlyName: 'App Event',
        ...SubComponentFields,
    },
    [exports.VideoConferencingKey]: {
        dir: exports.VideoConferencingKey,
        parentComponent: exports.AppKey,
        userFriendlyName: 'Video Conferencing',
        ...SubComponentFields,
    },
    [exports.WebhooksKey]: {
        dir: exports.WebhooksKey,
        parentComponent: exports.AppKey,
        userFriendlyName: 'Webhooks',
        ...SubComponentFields,
        singularComponent: true,
    },
    [exports.WebhooksJournalKey]: {
        dir: exports.WebhooksJournalKey,
        parentComponent: exports.AppKey,
        userFriendlyName: 'Webhooks Journal',
        ...SubComponentFields,
        singularComponent: true,
    },
    [exports.WorkflowActionsKey]: {
        dir: 'workflow-actions',
        parentComponent: exports.AppKey,
        userFriendlyName: 'Workflow Action',
        ...SubComponentFields,
    },
    [exports.TelemetryKey]: {
        dir: exports.TelemetryKey,
        parentComponent: exports.AppKey,
        userFriendlyName: 'Telemetry',
        singularComponent: true,
        ...SubComponentFields,
    },
    [exports.MCPRegistryKey]: {
        dir: exports.MCPRegistryKey,
        parentComponent: exports.AppKey,
        userFriendlyName: 'MCP Registry',
        singularComponent: true,
        ...SubComponentFields,
    },
};
exports.userFacingToInternalType = {
    [exports.AppKey]: 'APPLICATION',
    [exports.SCIMKey]: 'SCIM_INTEGRATION',
    [exports.TelemetryKey]: 'TELEMETRY_CONFIG',
    [exports.CmsAssetsKey]: 'REACT_THEME',
};
exports.internalTypeToUserFacing = Object.fromEntries(Object.entries(exports.userFacingToInternalType).map(([key, value]) => [value, key]));
exports.metafileExtension = '-hsmeta.json';
exports.profileFilePrefix = 'hsprofile';
exports.hsProjectJsonFilename = 'hsproject.json';
exports.packageJson = 'package.json';
exports.packageLockJson = 'package-lock.json';
function getSubComponentDirsForParentType(parentComponent) {
    return Object.values(exports.Components).reduce((acc, item) => {
        if (item.parentComponent === parentComponent) {
            acc.push(item.dir);
        }
        return acc;
    }, []);
}
exports.allowedAppSubComponentsDirs = getSubComponentDirsForParentType(exports.AppKey);
exports.allowedThemeSubComponentsDirs = getSubComponentDirsForParentType(exports.ThemeKey);
exports.allowedCmsAssetSubComponentDirs = getSubComponentDirsForParentType(exports.CmsAssetsKey);
exports.ProjectStructure = {
    [exports.AppKey]: exports.allowedAppSubComponentsDirs,
    [exports.ThemeKey]: exports.allowedThemeSubComponentsDirs,
    [exports.CmsAssetsKey]: exports.allowedCmsAssetSubComponentDirs,
};
exports.allowedComponentDirectories = Object.keys(exports.ProjectStructure);
exports.allowedSubComponentDirectories = Object.entries(exports.ProjectStructure)
    .map(([key, value]) => value.map((value) => path.join(key, value)))
    .flat();
// Profile variable types
exports.ProfileVariableTypes = {
    PROFILE_INT: 'PROFILE_INT',
    PROFILE_LONG: 'PROFILE_LONG',
    PROFILE_STRING: 'PROFILE_STRING',
    PROFILE_BOOLEAN: 'PROFILE_BOOLEAN',
};
exports.allowedTopLevelFields = new Set([
    'uid',
    'type',
    'config',
    'dependencies',
]);
