// @hubspot/project-parsing-lib/src/lib/migrateThemes.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProjectThemeDetails = getProjectThemeDetails;
exports.migrateThemes = migrateThemes;
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const util_1 = require("util");
const fs_2 = require("@hubspot/local-dev-lib/fs");
const constants_1 = require("./constants");
const utils_1 = require("./utils");
const copy_1 = require("../lang/copy");
const mkdtempAsync = (0, util_1.promisify)(fs_1.default.mkdtemp);
const LEGACY_THEME_CONFIG = 'theme.json';
const LEGACY_REACT_THEME_CONFIG = 'cms-assets.json';
async function getProjectThemeDetails(projectSourceDir) {
    const files = await (0, fs_2.walk)(projectSourceDir, ['node_modules']);
    const legacyThemeDetails = [];
    const legacyReactThemeDetails = [];
    files.forEach((filename) => {
        const isLegacyThemeConfig = filename.endsWith(LEGACY_THEME_CONFIG);
        const isLegacyReactThemeConfig = filename.endsWith(LEGACY_REACT_THEME_CONFIG);
        if (isLegacyThemeConfig || isLegacyReactThemeConfig) {
            const parsedConfig = (0, utils_1.loadJsonFile)(filename);
            const themeDetails = {
                configFilepath: filename,
                themePath: path_1.default.dirname(filename),
                themeConfig: parsedConfig,
            };
            if (isLegacyThemeConfig) {
                legacyThemeDetails.push(themeDetails);
            }
            else {
                legacyReactThemeDetails.push(themeDetails);
            }
        }
    });
    return {
        legacyThemeDetails,
        legacyReactThemeDetails,
    };
}
function buildNewTheme(componentKey, migratedProjectTempDir, themeDetails) {
    const themeName = path_1.default.basename(themeDetails.themePath);
    const componentDir = path_1.default.join(migratedProjectTempDir, componentKey);
    const newThemeDir = path_1.default.join(componentDir, themeName);
    fs_1.default.cpSync(themeDetails.themePath, newThemeDir, {
        recursive: true,
    });
    const newThemeConfig = {
        uid: themeName,
        type: componentKey,
        config: {
            themePath: themeName,
            secretNames: [],
        },
    };
    // Remove secrets from the legacy theme config if they exist & move them to the new theme config
    let updatedLegacyThemeConfig;
    if ('secret_names' in themeDetails.themeConfig) {
        const { secret_names, ...configWithoutSecrets } = themeDetails.themeConfig;
        if (secret_names) {
            newThemeConfig.config.secretNames = secret_names;
            updatedLegacyThemeConfig = configWithoutSecrets;
        }
    }
    else if ('secretNames' in themeDetails.themeConfig) {
        const { secretNames, ...configWithoutSecrets } = themeDetails.themeConfig;
        if (secretNames) {
            newThemeConfig.config.secretNames = secretNames;
            updatedLegacyThemeConfig = configWithoutSecrets;
        }
    }
    if (updatedLegacyThemeConfig) {
        const newLegacyConfigFilepath = path_1.default.join(newThemeDir, path_1.default.basename(themeDetails.configFilepath));
        fs_1.default.writeFileSync(newLegacyConfigFilepath, JSON.stringify(updatedLegacyThemeConfig, null, 2));
    }
    fs_1.default.writeFileSync(path_1.default.join(componentDir, `${themeName}${constants_1.metafileExtension}`), JSON.stringify(newThemeConfig, null, 2));
}
async function migrateThemes(projectDir, projectSourceDir) {
    const { legacyThemeDetails, legacyReactThemeDetails, } = await getProjectThemeDetails(projectSourceDir);
    // Migrate the project to a temporary directory to avoid file conflicts in the original project
    const migratedProjectTempDir = await mkdtempAsync(path_1.default.join(os_1.default.tmpdir(), 'hubspot-migrated-project-'));
    let isRootReactThemeProject = false;
    legacyReactThemeDetails.forEach(themeDetails => {
        isRootReactThemeProject = themeDetails.themePath === projectSourceDir;
        if (!isRootReactThemeProject) {
            buildNewTheme(constants_1.CmsAssetsKey, migratedProjectTempDir, themeDetails);
        }
    });
    // Do not support migrating themes that live at the root of the project src directory
    if (isRootReactThemeProject) {
        return {
            legacyThemeDetails,
            legacyReactThemeDetails,
            migrated: false,
            failureReason: copy_1.errorMessages.migrateThemes.rootReactThemeNotSupported,
        };
    }
    legacyThemeDetails.forEach(themeDetails => {
        buildNewTheme(constants_1.ThemeKey, migratedProjectTempDir, themeDetails);
    });
    // Copy the rest of the project files to the temp directory
    fs_1.default.cpSync(projectSourceDir, migratedProjectTempDir, {
        recursive: true,
        filter: src => {
            return (!legacyThemeDetails.some(file => src.includes(file.themePath)) &&
                !legacyReactThemeDetails.some(file => src.includes(file.themePath)));
        },
    });
    // Archive the legacy theme files
    const archiveDest = path_1.default.join(projectDir, 'archive');
    fs_1.default.renameSync(projectSourceDir, archiveDest);
    // Copy the new theme config files to the project source directory
    fs_1.default.cpSync(migratedProjectTempDir, projectSourceDir, {
        recursive: true,
    });
    fs_1.default.rmSync(migratedProjectTempDir, { recursive: true });
    return {
        legacyThemeDetails,
        legacyReactThemeDetails,
        migrated: true,
    };
}
