// @hubspot/project-parsing-lib/src/lib/profiles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getIsProfileFile = getIsProfileFile;
exports.getHsProfileFilename = getHsProfileFilename;
exports.getHsProfileName = getHsProfileName;
exports.getHsProfileVariables = getHsProfileVariables;
exports.applyHsProfileVariables = applyHsProfileVariables;
const constants_1 = require("./constants");
const copy_1 = require("../lang/copy");
const profileFilenameRegex = new RegExp(`^${constants_1.profileFilePrefix}\\.([^.]+)\\.json$`);
const profileInsertRegex = /\${(.*?)}/g;
function getIsProfileFile(filename) {
    return profileFilenameRegex.test(filename);
}
function getHsProfileFilename(profileName) {
    return `${constants_1.profileFilePrefix}.${profileName}.json`;
}
function getHsProfileName(profileFilename) {
    const match = profileFilename.match(profileFilenameRegex);
    return match ? match[1] : '';
}
function getHsProfileVariables(hsProfileContents) {
    const profileVariablesAreValid = hsProfileContents?.variables &&
        typeof hsProfileContents.variables === 'object' &&
        !Array.isArray(hsProfileContents.variables);
    return profileVariablesAreValid ? hsProfileContents.variables : {};
}
function interpolate(file, template, profileVariables) {
    return template.replace(profileInsertRegex, (__match, key) => {
        const profileVariableType = typeof profileVariables[key];
        if (profileVariableType === 'undefined') {
            throw new Error(copy_1.errorMessages.profile.missingValue(key, file));
        }
        if (!['string', 'number', 'boolean'].includes(profileVariableType)) {
            throw new Error(copy_1.errorMessages.profile.invalidValue(key));
        }
        return String(profileVariables[key]);
    });
}
function applyHsProfileVariables(fileParseResults, hsProfileContents) {
    const profileVariables = getHsProfileVariables(hsProfileContents);
    return fileParseResults.map(fileParseResult => {
        const { content, file } = fileParseResult;
        if (content && content.config) {
            let interpolatedConfig = content.config;
            interpolatedConfig = JSON.parse(interpolate(file, JSON.stringify(content.config), profileVariables));
            if (interpolatedConfig) {
                return {
                    ...fileParseResult,
                    content: {
                        ...content,
                        config: interpolatedConfig,
                    },
                };
            }
        }
        return fileParseResult;
    });
}
