// @hubspot/cli/commands/secret/listSecret.js
import { fetchSecrets } from '@hubspot/local-dev-lib/api/secrets';
import { logError, ApiErrorContext } from '../../lib/errorHandlers/index.js';
import { trackCommandUsage } from '../../lib/usageTracking.js';
import { uiAccountDescription } from '../../lib/ui/index.js';
import { commands } from '../../lang/en.js';
import { uiLogger } from '../../lib/ui/logger.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
const command = 'list';
const describe = commands.secret.subcommands.list.describe;
async function handler(args) {
    const { derivedAccountId } = args;
    trackCommandUsage('secrets-list', {}, derivedAccountId);
    try {
        const { data: { results }, } = await fetchSecrets(derivedAccountId);
        const groupLabel = commands.secret.subcommands.list.groupLabel(uiAccountDescription(derivedAccountId));
        uiLogger.group(groupLabel);
        results.forEach(secret => uiLogger.log(secret));
        uiLogger.groupEnd();
    }
    catch (err) {
        uiLogger.error(commands.secret.subcommands.list.errors.list);
        logError(err, new ApiErrorContext({
            request: 'add secret',
            accountId: derivedAccountId,
        }));
    }
}
function listSecretBuilder(yargs) {
    return yargs;
}
const builder = makeYargsBuilder(listSecretBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const listSecretCommand = {
    command,
    describe,
    handler,
    builder,
};
export default listSecretCommand;
