// @hubspot/cli/ui/components/StatusMessageBoxes.js
import { jsx as _jsx } from "react/jsx-runtime";
import { INK_COLORS } from '../styles.js';
import { BoxWithTitle } from './BoxWithTitle.js';
export function getWarningBox(props) {
    return _jsx(WarningBox, { ...props });
}
export function WarningBox({ title, message, textCentered, }) {
    const color = INK_COLORS.WARNING_YELLOW;
    return (_jsx(BoxWithTitle, { title: title, message: message, titleBackgroundColor: color, borderColor: color, textCentered: textCentered }));
}
export function getAlertBox(props) {
    return _jsx(AlertBox, { ...props });
}
export function AlertBox({ title, message, textCentered, }) {
    const color = INK_COLORS.ALERT_RED;
    return (_jsx(BoxWithTitle, { title: title, message: message, titleBackgroundColor: color, borderColor: color, textCentered: textCentered }));
}
export function getSuccessBox(props) {
    return _jsx(SuccessBox, { ...props });
}
export function SuccessBox({ title, message, textCentered, }) {
    const color = INK_COLORS.SUCCESS_GREEN;
    return (_jsx(BoxWithTitle, { title: title, message: message, titleBackgroundColor: color, borderColor: color, textCentered: textCentered }));
}
export function getInfoBox(props) {
    return _jsx(InfoBox, { ...props });
}
export function InfoBox({ title, message, textCentered, }) {
    const color = INK_COLORS.INFO_BLUE;
    return (_jsx(BoxWithTitle, { title: title, message: message, titleBackgroundColor: color, borderColor: color, textCentered: textCentered }));
}
