// @hubspot/cli/lib/ui/index.js
import chalk from 'chalk';
import { getConfigAccountIfExists, globalConfigFileExists, } from '@hubspot/local-dev-lib/config';
import { uiLogger } from './logger.js';
import { supportsHyperlinkModule } from './supportHyperlinks.js';
import { supportsColor } from './supportsColor.js';
import { uiMessages } from './uiMessages.js';
import { HUBSPOT_ACCOUNT_TYPE_STRINGS } from '@hubspot/local-dev-lib/constants/config';
export const UI_COLORS = {
    SORBET: '#FF8F59',
    MARIGOLD: '#f5c26b',
    MARIGOLD_DARK: '#dbae60',
};
export function uiLine() {
    uiLogger.log('-'.repeat(50));
}
export function getTerminalUISupport() {
    return {
        hyperlinks: supportsHyperlinkModule.stdout,
        color: supportsColor.stdout.hasBasic,
    };
}
export function uiLink(linkText, url) {
    const terminalUISupport = getTerminalUISupport();
    const encodedUrl = encodeURI(url);
    if (terminalUISupport.hyperlinks) {
        const CLOSE_SEQUENCE = '\u001B]8;;\u0007';
        const result = [
            '\u001B]8;;',
            encodedUrl,
            '\u0007',
            linkText,
            CLOSE_SEQUENCE,
        ].join('');
        return terminalUISupport.color ? chalk.cyan(result) : result;
    }
    return terminalUISupport.color
        ? `${linkText}: ${chalk.reset.cyan(encodedUrl)}`
        : `${linkText}: ${encodedUrl}`;
}
export function uiAccountDescription(accountId, bold = true) {
    const account = accountId ? getConfigAccountIfExists(accountId) : undefined;
    let message;
    if (account && account.accountType) {
        message = `${account.name ? `${account.name} ` : ''}[${HUBSPOT_ACCOUNT_TYPE_STRINGS[account.accountType]}] (${accountId})`;
    }
    else {
        message = accountId ? accountId.toString() : '';
    }
    if (message.trim().length === 0) {
        return 'unknown account';
    }
    return bold ? chalk.bold(message) : message;
}
export function uiInfoSection(title, logContent) {
    uiLine();
    uiLogger.log(chalk.bold(title));
    uiLogger.log('');
    logContent();
    uiLogger.log('');
    uiLine();
}
export function uiCommandReference(command, withQuotes = true) {
    const terminalUISupport = getTerminalUISupport();
    const commandReference = withQuotes ? `\`${command}\`` : command;
    return chalk.bold(terminalUISupport.color
        ? chalk.hex(UI_COLORS.MARIGOLD_DARK)(commandReference)
        : commandReference);
}
export function uiAuthCommandReference({ accountId, qa, } = {}) {
    const userIsUsingGlobalConfig = globalConfigFileExists();
    let command = 'hs auth';
    if (userIsUsingGlobalConfig) {
        command = 'hs account auth';
    }
    return uiCommandReference(`${command}${accountId ? ` --account=${accountId}` : ''}${qa ? ' --qa' : ''}`);
}
export function uiFeatureHighlight(features, title) {
    uiInfoSection(title ? title : uiMessages.featureHighlight.defaultTitle, () => {
        features.forEach(feature => {
            const featureConfig = uiMessages.featureHighlight.featureKeys[feature];
            if (!featureConfig) {
                uiLogger.debug(`Feature config not found for: ${feature}`);
                return;
            }
            let message;
            if ('linkText' in featureConfig && 'url' in featureConfig) {
                // linkText + url (for sampleProjects)
                message = featureConfig.message(uiLink(featureConfig.linkText, featureConfig.url));
            }
            else if ('command' in featureConfig && 'message' in featureConfig) {
                // Command + Message function (most cases)
                message = featureConfig.message(uiCommandReference(featureConfig.command));
            }
            else {
                // Message only (for projectCommandTip)
                message = featureConfig.message;
            }
            uiLogger.log(`  - ${message}`);
        });
    });
}
export function uiBetaTag(message, log = true) {
    const tag = uiMessages.betaTag;
    const result = `${tag} ${message}`;
    if (log) {
        uiLogger.log(result);
        return;
    }
    return result;
}
export function uiDeprecatedTag(message, log = true) {
    const tag = uiMessages.deprecatedTag;
    const result = `${tag} ${message}`;
    if (log) {
        uiLogger.log(result);
        return;
    }
    return result;
}
export function uiCommandDisabledBanner(command, url, message) {
    const tag = message ||
        `The ${uiCommandReference(command)} command is disabled. Run ${uiCommandReference('npm i -g @hubspot/cli@latest')} to update to the latest HubSpot CLI version. ${url ? uiLink('See all HubSpot CLI commands here.', url) : ''}`;
    uiLogger.log('');
    uiLine();
    uiLogger.error(tag);
    uiLine();
    uiLogger.log('');
}
export function uiDeprecatedDescription(message, command, url) {
    const tag = message || uiMessages.disabledMessage(command, url);
    return uiDeprecatedTag(tag);
}
export function uiCommandRenamedDescription(describe, newCommand) {
    return uiDeprecatedTag(`${describe} ${uiMessages.commandRenamedMessage(newCommand)}`, false);
}
export function uiDeprecatedMessage(command, url, message) {
    const tag = uiMessages.deprecatedDescription(message || '', command, url);
    uiLogger.log('');
    uiDeprecatedTag(tag);
    uiLogger.log('');
}
export function uiCommandRelocatedMessage(newCommand) {
    uiLogger.log('');
    uiDeprecatedTag(uiMessages.commandRenamedMessage(newCommand));
    uiLogger.log('');
}
export function indent(level) {
    const indentation = '  ';
    return indentation.repeat(level);
}
