// @hubspot/cli/lib/ui/SpinniesManager.js
/*
https://github.com/jbcarpanelli/spinnies

Copyright 2019 Juan Bautista Carpanelli (jcarpanelli)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
**/
import readline from 'readline';
import chalk from 'chalk';
import cliCursor from 'cli-cursor';
import { breakText, cleanStream, colorOptions, getLinesLength, prefixOptions, purgeSpinnerOptions, purgeSpinnersOptions, SPINNERS, terminalSupportsUnicode, writeStream, } from './spinniesUtils.js';
import { uiLogger } from './logger.js';
function safeColor(text, color) {
    const chalkFn = chalk[color];
    if (typeof chalkFn === 'function') {
        return chalkFn(text);
    }
    return text;
}
class SpinniesManager {
    disableOutput = false;
    options;
    spinners = {};
    isCursorHidden = false;
    currentInterval = null;
    stream = process.stderr;
    lineCount = 0;
    currentFrameIndex = 0;
    spin;
    constructor() {
        this.resetState();
    }
    init(options = {}) {
        this.options = {
            spinnerColor: 'greenBright',
            succeedColor: 'green',
            failColor: 'red',
            spinner: terminalSupportsUnicode() ? SPINNERS.dots : SPINNERS.dashes,
            disableSpins: false,
            ...purgeSpinnersOptions(options),
        };
        this.spin =
            !this.options.disableSpins &&
                !process.env.CI &&
                process.stderr &&
                process.stderr.isTTY;
        if (!this.hasAnySpinners()) {
            this.resetState();
        }
        this.bindSigint();
    }
    resetState() {
        // Default Spinnies fields
        this.spinners = {};
        this.isCursorHidden = false;
        if (this.currentInterval) {
            clearInterval(this.currentInterval);
        }
        this.currentInterval = null;
        this.stream = process.stderr;
        this.lineCount = 0;
        this.currentFrameIndex = 0;
    }
    setDisableOutput(disableOutput) {
        this.disableOutput = disableOutput;
    }
    pick(name) {
        return this.spinners?.[name];
    }
    add(name, options = {}) {
        // Support adding generic spinnies lines without specifying a name
        const resolvedName = name || `${Date.now()}-${Math.random()}`;
        if (!options.text) {
            options.text = resolvedName;
        }
        this.spinners[resolvedName] = {
            ...colorOptions(this.options),
            succeedPrefix: this.options.succeedPrefix,
            failPrefix: this.options.failPrefix,
            status: 'spinning',
            ...purgeSpinnerOptions(options),
        };
        this.updateSpinnerState();
    }
    update(name, options = {}) {
        const { status } = options;
        this.setSpinnerProperties(name, options, status);
        this.updateSpinnerState();
    }
    succeed(name, options = {}) {
        this.setSpinnerProperties(name, options, 'succeed');
        this.updateSpinnerState();
    }
    fail(name, options = {}) {
        this.setSpinnerProperties(name, options, 'fail');
        this.updateSpinnerState();
    }
    remove(name) {
        if (typeof name !== 'string') {
            uiLogger.debug('A spinner reference name must be specified');
            return;
        }
        delete this.spinners[name];
    }
    stopAll(newStatus = 'stopped') {
        Object.keys(this.spinners).forEach(name => {
            const { status: currentStatus } = this.spinners[name];
            if (currentStatus !== 'fail' &&
                currentStatus !== 'succeed' &&
                currentStatus !== 'non-spinnable') {
                if (newStatus === 'succeed' || newStatus === 'fail') {
                    this.spinners[name].status = newStatus;
                    this.spinners[name].color = this.options[`${newStatus}Color`];
                }
                else {
                    this.spinners[name].status = 'stopped';
                    this.spinners[name].color = 'gray';
                }
            }
        });
        this.checkIfActiveSpinners();
        return this.spinners;
    }
    hasAnySpinners() {
        return !!Object.keys(this.spinners).length;
    }
    hasActiveSpinners() {
        return !!Object.values(this.spinners).find(({ status }) => status === 'spinning');
    }
    setSpinnerProperties(name, options, status) {
        if (typeof name !== 'string') {
            uiLogger.debug('A spinner reference name must be specified');
            return;
        }
        if (!this.spinners[name]) {
            uiLogger.debug(`No spinner initialized with name ${name}`);
            return;
        }
        options = purgeSpinnerOptions(options);
        status = status || 'spinning';
        this.spinners[name] = { ...this.spinners[name], ...options, status };
    }
    updateSpinnerState() {
        if (this.disableOutput) {
            return;
        }
        if (this.spin) {
            if (this.currentInterval) {
                clearInterval(this.currentInterval);
            }
            this.currentInterval = this.loopStream();
            if (!this.isCursorHidden) {
                cliCursor.hide();
            }
            this.isCursorHidden = true;
            this.checkIfActiveSpinners();
        }
        else {
            this.setRawStreamOutput();
        }
    }
    loopStream() {
        const frames = this.options.spinner?.frames || SPINNERS.dots.frames;
        const interval = this.options.spinner?.interval || SPINNERS.dots.interval;
        return setInterval(() => {
            this.setStreamOutput(frames[this.currentFrameIndex]);
            this.currentFrameIndex =
                this.currentFrameIndex === frames.length - 1
                    ? 0
                    : ++this.currentFrameIndex;
        }, interval);
    }
    setStreamOutput(frame = '') {
        let output = '';
        const linesLength = [];
        const hasActiveSpinners = this.hasActiveSpinners();
        Object.values(this.spinners).forEach(spinner => {
            let { text } = spinner;
            const { status, color, spinnerColor, succeedColor, failColor, indent = 0, succeedPrefix = prefixOptions(this.options).succeedPrefix, failPrefix = prefixOptions(this.options).failPrefix, } = spinner;
            let line;
            let prefixLength = indent;
            text = text ?? '';
            if (status === 'spinning') {
                prefixLength += frame.length + 1;
                text = breakText(text, prefixLength);
                const colorizedFrame = safeColor(frame, spinnerColor);
                const colorizedText = safeColor(text, color);
                line = `${colorizedFrame} ${colorizedText}`;
            }
            else {
                if (status === 'succeed') {
                    prefixLength += succeedPrefix.length + 1;
                    if (hasActiveSpinners) {
                        text = breakText(text, prefixLength);
                    }
                    const colorizedText = safeColor(text, succeedColor);
                    line = `${chalk.green(succeedPrefix)} ${colorizedText}`;
                }
                else if (status === 'fail') {
                    prefixLength += failPrefix.length + 1;
                    if (hasActiveSpinners) {
                        text = breakText(text, prefixLength);
                    }
                    const colorizedText = safeColor(text, failColor);
                    line = `${chalk.red(failPrefix)} ${colorizedText}`;
                }
                else {
                    if (hasActiveSpinners) {
                        text = breakText(text, prefixLength);
                    }
                    line = safeColor(text, color);
                }
            }
            linesLength.push(...getLinesLength(text, prefixLength));
            output += indent ? `${' '.repeat(indent)}${line}\n` : `${line}\n`;
        });
        if (!hasActiveSpinners) {
            readline.clearScreenDown(this.stream);
        }
        writeStream(this.stream, output, linesLength);
        if (hasActiveSpinners) {
            cleanStream(this.stream, linesLength);
        }
        this.lineCount = linesLength.length;
    }
    setRawStreamOutput() {
        Object.values(this.spinners).forEach(i => {
            process.stderr.write(`- ${i.text}\n`);
        });
    }
    checkIfActiveSpinners() {
        if (!this.hasActiveSpinners()) {
            if (this.spin) {
                this.setStreamOutput();
                readline.moveCursor(this.stream, 0, this.lineCount);
                if (this.currentInterval) {
                    clearInterval(this.currentInterval);
                }
                this.isCursorHidden = false;
                cliCursor.show();
            }
            this.spinners = {};
        }
    }
    bindSigint() {
        process.removeAllListeners('SIGINT');
        process.on('SIGINT', () => {
            cliCursor.show();
            readline.moveCursor(process.stderr, 0, this.lineCount);
            process.exit(0);
        });
    }
}
const toExport = new SpinniesManager();
export default toExport;
