// @hubspot/cli/commands/cms/function/deploy.js
import { uiLogger } from '../../../lib/ui/logger.js';
import { buildPackage, getBuildStatus, } from '@hubspot/local-dev-lib/api/functions';
import { isHubSpotHttpError } from '@hubspot/local-dev-lib/errors/index';
import SpinniesManager from '../../../lib/ui/SpinniesManager.js';
import { trackCommandUsage } from '../../../lib/usageTracking.js';
import { logError, ApiErrorContext } from '../../../lib/errorHandlers/index.js';
import { uiAccountDescription } from '../../../lib/ui/index.js';
import { poll } from '../../../lib/polling.js';
import { outputBuildLog } from '../../../lib/serverlessLogs.js';
import { commands } from '../../../lang/en.js';
import { makeYargsBuilder } from '../../../lib/yargsUtils.js';
function isFunctionBuildError(e) {
    return (typeof e === 'object' && e !== null && 'status' in e && e.status === 'ERROR');
}
const command = 'deploy <path>';
const describe = undefined; // Hidden command
async function handler(args) {
    const { path: functionPath, derivedAccountId } = args;
    const splitFunctionPath = functionPath.split('.');
    trackCommandUsage('function-deploy', undefined, derivedAccountId);
    if (!splitFunctionPath.length ||
        splitFunctionPath[splitFunctionPath.length - 1] !== 'functions') {
        uiLogger.error(commands.cms.subcommands.function.subcommands.deploy.errors.notFunctionsFolder(functionPath));
        return;
    }
    uiLogger.debug(commands.cms.subcommands.function.subcommands.deploy.debug.startingBuildAndDeploy(functionPath));
    SpinniesManager.init();
    SpinniesManager.add('loading', {
        text: commands.cms.subcommands.function.subcommands.deploy.loading(functionPath, uiAccountDescription(derivedAccountId)),
    });
    try {
        const { data: buildId } = await buildPackage(derivedAccountId, functionPath);
        const successResp = await poll(() => getBuildStatus(derivedAccountId, Number(buildId)));
        if (successResp) {
            const buildTimeSeconds = successResp.buildTime
                ? (successResp.buildTime / 1000).toFixed(2)
                : 0;
            SpinniesManager.succeed('loading');
            if (successResp.cdnUrl) {
                await outputBuildLog(successResp.cdnUrl);
            }
            uiLogger.success(commands.cms.subcommands.function.subcommands.deploy.success.deployed(functionPath, derivedAccountId, buildTimeSeconds));
        }
    }
    catch (e) {
        SpinniesManager.fail('loading', {
            text: commands.cms.subcommands.function.subcommands.deploy.loadingFailed(functionPath, uiAccountDescription(derivedAccountId)),
        });
        if (isHubSpotHttpError(e) && e.status === 404) {
            uiLogger.error(commands.cms.subcommands.function.subcommands.deploy.errors.noPackageJson(functionPath));
        }
        else if (isFunctionBuildError(e)) {
            await outputBuildLog(e.cdnUrl);
            uiLogger.error(commands.cms.subcommands.function.subcommands.deploy.errors.buildError(String(e.errorReason)));
        }
        else {
            logError(e, new ApiErrorContext({
                accountId: derivedAccountId,
                request: functionPath,
            }));
        }
    }
}
function functionDeployBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.function.subcommands.deploy.positionals.path
            .describe,
        type: 'string',
    });
    yargs.example([
        [
            '$0 functions deploy myFunctionFolder.functions',
            commands.cms.subcommands.function.subcommands.deploy.examples.default,
        ],
    ]);
    return yargs;
}
const builder = makeYargsBuilder(functionDeployBuilder, command, describe, {
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const functionDeployCommand = {
    command,
    describe,
    handler,
    builder,
};
export default functionDeployCommand;
