// @hubspot/cli/commands/cms/function/server.js
import { trackCommandUsage } from '../../../lib/usageTracking.js';
import { uiLogger } from '../../../lib/ui/logger.js';
// This package is not typed, so we need to use require
import { start as startTestServer } from '@hubspot/serverless-dev-runtime';
import { commands } from '../../../lang/en.js';
import { makeYargsBuilder } from '../../../lib/yargsUtils.js';
const command = 'server <path>';
const describe = undefined;
async function handler(args) {
    const { path: functionPath, derivedAccountId } = args;
    trackCommandUsage('functions-server', undefined, derivedAccountId);
    uiLogger.debug(commands.cms.subcommands.function.subcommands.server.debug.startingServer(functionPath));
    startTestServer({
        accountId: derivedAccountId,
        ...args,
    });
}
function functionServerBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.function.subcommands.server.positionals.path
            .describe,
        type: 'string',
    });
    yargs.options({
        port: {
            describe: commands.cms.subcommands.function.subcommands.server.options.port
                .describe,
            type: 'string',
            default: 5432,
        },
        contact: {
            describe: commands.cms.subcommands.function.subcommands.server.options.contact
                .describe,
            type: 'boolean',
            default: true,
        },
        watch: {
            describe: commands.cms.subcommands.function.subcommands.server.options.watch
                .describe,
            type: 'boolean',
            default: true,
        },
        'log-output': {
            describe: commands.cms.subcommands.function.subcommands.server.options.logOutput
                .describe,
            type: 'boolean',
            default: false,
        },
    });
    yargs.example([
        [
            '$0 functions server ./tmp/myFunctionFolder.functions',
            commands.cms.subcommands.function.subcommands.server.examples.default,
        ],
    ]);
    return yargs;
}
const builder = makeYargsBuilder(functionServerBuilder, command, describe, {
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const functionServerCommand = {
    command,
    describe,
    handler,
    builder,
};
export default functionServerCommand;
