// @hubspot/cli/lib/accountTypes.js
import { HUBSPOT_ACCOUNT_TYPES } from '@hubspot/local-dev-lib/constants/config';
import { hasUnfiedAppsAccess } from './hasFeature.js';
function isAccountType(accountConfig, accountType) {
    return Boolean(accountConfig.accountType && accountType.includes(accountConfig.accountType));
}
export function isStandardAccount(accountConfig) {
    return isAccountType(accountConfig, [HUBSPOT_ACCOUNT_TYPES.STANDARD]);
}
export function isSandbox(accountConfig) {
    return isAccountType(accountConfig, [
        HUBSPOT_ACCOUNT_TYPES.STANDARD_SANDBOX,
        HUBSPOT_ACCOUNT_TYPES.DEVELOPMENT_SANDBOX,
    ]);
}
export function isStandardSandbox(accountConfig) {
    return isAccountType(accountConfig, [HUBSPOT_ACCOUNT_TYPES.STANDARD_SANDBOX]);
}
export function isDevelopmentSandbox(accountConfig) {
    return isAccountType(accountConfig, [
        HUBSPOT_ACCOUNT_TYPES.DEVELOPMENT_SANDBOX,
    ]);
}
export function isDeveloperTestAccount(accountConfig) {
    return isAccountType(accountConfig, [HUBSPOT_ACCOUNT_TYPES.DEVELOPER_TEST]);
}
export function isAppDeveloperAccount(accountConfig) {
    return isAccountType(accountConfig, [HUBSPOT_ACCOUNT_TYPES.APP_DEVELOPER]);
}
export function isTestAccountOrSandbox(accountConfig) {
    return isAccountType(accountConfig, [
        HUBSPOT_ACCOUNT_TYPES.DEVELOPER_TEST,
        HUBSPOT_ACCOUNT_TYPES.STANDARD_SANDBOX,
        HUBSPOT_ACCOUNT_TYPES.DEVELOPMENT_SANDBOX,
    ]);
}
export async function isUnifiedAccount(account) {
    if (!account.accountId) {
        return false;
    }
    return hasUnfiedAppsAccess(account.accountId);
}
