// @hubspot/cli/lib/links.js
import open from 'open';
import { getConfigAccountEnvironment } from '@hubspot/local-dev-lib/config';
import { getHubSpotWebsiteOrigin } from '@hubspot/local-dev-lib/urls';
import { ENVIRONMENTS } from '@hubspot/local-dev-lib/constants/environments';
import { uiLogger } from './ui/logger.js';
import { getTableContents, getTableHeader } from './ui/table.js';
const COMMON_SITE_LINKS = {
    APPS_MARKETPLACE: {
        shortcut: 'apps-marketplace',
        alias: 'apm',
        getUrl: (accountId, baseUrl) => `${baseUrl}/ecosystem/${accountId}/marketplace/apps`,
    },
    ASSET_MARKETPLACE: {
        shortcut: 'asset-marketplace',
        alias: 'asm',
        getUrl: (accountId, baseUrl) => `${baseUrl}/ecosystem/${accountId}/marketplace/products`,
    },
    CONTENT_STAGING: {
        shortcut: 'content-staging',
        alias: 'cs',
        getUrl: (accountId, baseUrl) => `${baseUrl}/content/${accountId}/staging`,
    },
    DESIGN_MANAGER: {
        shortcut: 'design-manager',
        alias: 'dm',
        getUrl: (accountId, baseUrl) => `${baseUrl}/design-manager/${accountId}`,
    },
    DOCS: {
        shortcut: 'docs',
        getUrl: () => 'https://developers.hubspot.com',
    },
    FILE_MANAGER: {
        shortcut: 'file-manager',
        alias: 'fm',
        getUrl: (accountId, baseUrl) => `${baseUrl}/files/${accountId}`,
    },
    FORUMS: {
        shortcut: 'forums',
        getUrl: () => 'https://community.hubspot.com',
    },
    HUBDB: {
        shortcut: 'hubdb',
        alias: 'hdb',
        getUrl: (accountId, baseUrl) => `${baseUrl}/hubdb/${accountId}`,
    },
    SETTINGS: {
        shortcut: 'settings',
        alias: 's',
        getUrl: (accountId, baseUrl) => `${baseUrl}/settings/${accountId}`,
    },
    SETTINGS_NAVIGATION: {
        shortcut: 'settings/navigation',
        alias: 'sn',
        getUrl: (accountId, baseUrl) => `${baseUrl}/menus/${accountId}/edit/`,
    },
    SETTINGS_WEBSITE: {
        shortcut: 'settings/website',
        alias: 'sw',
        getUrl: (accountId, baseUrl) => `${baseUrl}/settings/${accountId}/website/pages/all-domains/page-templates`,
    },
    SETTINGS_getUrl_REDIRECTS: {
        shortcut: 'settings/url-redirects',
        alias: 'sur',
        getUrl: (accountId, baseUrl) => `${baseUrl}/domains/${accountId}/url-redirects`,
    },
    PURCHASED_ASSETS: {
        shortcut: 'purchased-assets',
        alias: 'pa',
        getUrl: (accountId, baseUrl) => `${baseUrl}/marketplace/${accountId}/manage-purchases`,
    },
    WEBSITE_PAGES: {
        shortcut: 'website-pages',
        alias: 'wp',
        getUrl: (accountId, baseUrl) => `${baseUrl}/website/${accountId}/pages/site`,
    },
};
export function getSiteLinksAsArray(accountId) {
    const baseUrl = getHubSpotWebsiteOrigin(getConfigAccountEnvironment(accountId));
    return Object.values(COMMON_SITE_LINKS)
        .sort((a, b) => (a.shortcut < b.shortcut ? -1 : 1))
        .map(l => ({ ...l, url: l.getUrl(accountId, baseUrl) }));
}
export function logSiteLinks(accountId) {
    const linksAsArray = getSiteLinksAsArray(accountId).map(l => [
        `${l.shortcut}${l.alias ? ` [alias: ${l.alias}]` : ''}`,
        '=>',
        l.url,
    ]);
    linksAsArray.unshift(getTableHeader(['Shortcut', '', 'Url']));
    uiLogger.log(getTableContents(linksAsArray));
}
export function openLink(accountId, shortcut) {
    const match = Object.values(COMMON_SITE_LINKS).find(l => l.shortcut === shortcut || (l.alias && l.alias === shortcut));
    if (!match) {
        uiLogger.error(`We couldn't find a shortcut matching ${shortcut}.  Type 'hs open list' to see a list of available shortcuts`);
        return;
    }
    const baseUrl = getHubSpotWebsiteOrigin(getConfigAccountEnvironment(accountId));
    open(match.getUrl(accountId, baseUrl), { url: true });
    uiLogger.success(`We opened ${match.getUrl(accountId, baseUrl)} in your browser`);
}
export function getProductUpdatesUrl(rolloutId, accountId) {
    const baseUrl = getHubSpotWebsiteOrigin(accountId ? getConfigAccountEnvironment(accountId) : ENVIRONMENTS.PROD);
    if (accountId) {
        return `${baseUrl}/product-updates/${accountId}/in-beta?rollout=${rolloutId}`;
    }
    return `${baseUrl}/l/product-updates/in-beta?rollout=${rolloutId}`;
}
