// @hubspot/cli/lib/usageTracking.js
import { trackUsage } from '@hubspot/local-dev-lib/trackUsage';
import { getConfigAccountById, getConfig } from '@hubspot/local-dev-lib/config';
import { API_KEY_AUTH_METHOD } from '@hubspot/local-dev-lib/constants/auth';
import { uiLogger } from './ui/logger.js';
import { pkg } from './jsonLoader.js';
import { debugError } from './errorHandlers/index.js';
const version = pkg.version;
export const EventClass = {
    USAGE: 'USAGE',
    INTERACTION: 'INTERACTION',
    VIEW: 'VIEW',
    ACTIVATION: 'ACTIVATION',
};
export function getNodeVersionData() {
    return {
        nodeVersion: process.version,
        nodeMajorVersion: (process.version || '').split('.')[0],
    };
}
export function getPlatform() {
    switch (process.platform) {
        case 'darwin':
            return 'macos';
        case 'win32':
            return 'windows';
        default:
            return process.platform;
    }
}
export async function trackCommandUsage(command, meta = {}, accountId) {
    try {
        const config = getConfig();
        if (config?.allowUsageTracking === false) {
            return;
        }
    }
    catch (e) { }
    uiLogger.debug(`Attempting to track usage of "${command}" command`);
    let authType = 'unknown';
    if (accountId) {
        const accountConfig = getConfigAccountById(accountId);
        authType =
            accountConfig && accountConfig.authType
                ? accountConfig.authType
                : API_KEY_AUTH_METHOD.value;
    }
    return trackCliInteraction({
        action: 'cli-command',
        command,
        authType,
        meta,
        accountId,
    });
}
export async function trackHelpUsage(command) {
    const config = getConfig();
    if (config?.allowUsageTracking === false) {
        return;
    }
    if (command) {
        uiLogger.debug(`Tracking help usage of "${command}" sub-command`);
    }
    else {
        uiLogger.debug('Tracking help usage of main command');
    }
    return trackCliInteraction({
        action: 'cli-help',
        command,
    });
}
export async function trackConvertFieldsUsage(command) {
    return trackCliInteraction({
        action: 'cli-process-fields',
        command,
    });
}
export async function trackAuthAction(command, authType, step, accountId) {
    return trackCliInteraction({
        action: 'cli-auth',
        command,
        authType,
        accountId,
        meta: {
            step,
        },
    });
}
export async function trackCommandMetadataUsage(command, meta = {}, accountId) {
    const config = getConfig();
    if (config?.allowUsageTracking === false) {
        return;
    }
    uiLogger.debug(`Attempting to track metadata usage of "${command}" command`);
    let authType = 'unknown';
    if (accountId) {
        const accountConfig = getConfigAccountById(accountId);
        authType =
            accountConfig && accountConfig.authType
                ? accountConfig.authType
                : API_KEY_AUTH_METHOD.value;
    }
    return trackCliInteraction({
        action: 'cli-command-metadata',
        command,
        authType,
        accountId,
        meta,
    });
}
async function trackCliInteraction({ action, accountId, command, authType, meta = {}, }) {
    try {
        const config = getConfig();
        if (config?.allowUsageTracking === false) {
            return;
        }
        const usageTrackingEvent = {
            action,
            os: getPlatform(),
            ...getNodeVersionData(),
            version,
            command,
            authType,
            ...meta,
        };
        if (process.env.HUBSPOT_MCP_AI_AGENT) {
            try {
                await trackUsage('cli-interaction', EventClass.INTERACTION, {
                    ...usageTrackingEvent,
                    action: 'cli-mcp-server',
                    type: process.env.HUBSPOT_MCP_AI_AGENT,
                }, accountId);
                uiLogger.debug('Sent AI usage tracking command event:', {
                    ...usageTrackingEvent,
                    action: 'cli-mcp-server',
                    type: process.env.HUBSPOT_MCP_AI_AGENT,
                });
            }
            catch (error) {
                debugError(error);
            }
        }
        try {
            uiLogger.debug('Sent usage tracking command event:', usageTrackingEvent);
            return trackUsage('cli-interaction', EventClass.INTERACTION, usageTrackingEvent, accountId);
        }
        catch (error) {
            debugError(error);
        }
    }
    catch (e) {
        debugError(e);
    }
}
