// @hubspot/ui-extensions-dev-server/dist/lib/plugins/codeBlockingPlugin.js
import { isNodeModule } from "../utils.js";
import { traverseAbstractSyntaxTree } from "../ast.js";
const codeBlockingPlugin = ({ logger, extensionPath }) => {
    return {
        name: 'ui-extensions-code-blocking-plugin',
        enforce: 'post', // run after default rollup plugins
        transform(code, filename) {
            if (isNodeModule(filename)) {
                return { code, map: null }; // We don't want to parse node modules
            }
            let sourceCodeMetadata = {
                functions: {},
                badImports: [],
                dataDependencies: {
                    importedHooks: {},
                    dependencies: [],
                },
                variableDeclarations: new Map(),
            };
            const requireFunctionName = 'require';
            try {
                //  Not sure why the types don't match for this.parse and the Rollup docs, but
                // the docs over on rollup's site specify ESTree.Program as the return type,
                // and the underlying data matches that https://rollupjs.org/plugin-development/#this-parse
                const abstractSyntaxTree = this.parse(code);
                sourceCodeMetadata = traverseAbstractSyntaxTree(abstractSyntaxTree, [{ functionName: requireFunctionName }], extensionPath, logger);
                if (sourceCodeMetadata.badImports) {
                    for (const badImport of sourceCodeMetadata.badImports) {
                        logger.warn(`Importing files from outside of the extension directory is not supported. Please move the import ${badImport} into the extension directory.`);
                    }
                }
            }
            catch (e) {
                logger.debug('Unable to parse and traverse source code');
                return { code, map: null };
            }
            if (sourceCodeMetadata.functions[requireFunctionName] &&
                sourceCodeMetadata.functions[requireFunctionName].scope === 'Global') {
                logger.warn('require statements are not supported, replace require statements with import');
            }
            return { code, map: null };
        },
    };
};
export default codeBlockingPlugin;
