// @hubspot/ui-extensions-dev-server/dist/lib/plugins/friendlyLoggingPlugin.js
import path from 'path';
const unfriendlyCodeMapper = Object.freeze({
    UNRESOLVED_IMPORT: (loggable) => {
        const { exporter, id } = loggable;
        const { base: extension } = path.parse(id);
        return {
            message: `${exporter} is imported by ${extension}, but ${exporter} cannot be resolved. Make sure ${exporter} is installed.`,
            level: 'error',
        };
    },
    MISSING_GLOBAL_NAME: () => {
        return {};
    },
});
const unfriendlyCodeList = Object.freeze(Object.keys(unfriendlyCodeMapper));
function friendlyLoggingPlugin({ logger }) {
    return {
        name: 'ui-extensions-friendly-logging-plugin',
        enforce: 'post',
        onLog(_level, log) {
            if (unfriendlyCodeList.includes(log.code || '')) {
                const { message, level } = _mapMessageToFriendlyVersion(log);
                if (message && level) {
                    logger[level](message);
                }
                return false; // Filter the log message
            }
            return true; // We don't have a friendly log message, let it through
        },
    };
}
function _mapMessageToFriendlyVersion(loggable) {
    const { code } = loggable;
    return unfriendlyCodeMapper[code]?.(loggable) ?? {};
}
export default friendlyLoggingPlugin;
