// @hubspot/cli/commands/cms/theme/create.js
import fs from 'fs-extra';
import { logError } from '../../../lib/errorHandlers/index.js';
import { resolveLocalPath } from '../../../lib/filesystem.js';
import { trackCommandUsage } from '../../../lib/usageTracking.js';
import { commands } from '../../../lang/en.js';
import { uiLogger } from '../../../lib/ui/logger.js';
import { makeYargsBuilder } from '../../../lib/yargsUtils.js';
import { EXIT_CODES } from '../../../lib/enums/exitCodes.js';
import assets from '../../../lib/cmsAssets/index.js';
const command = 'create';
const describe = commands.cms.subcommands.theme.subcommands.create.describe;
async function handler(args) {
    const { derivedAccountId, dest } = args;
    const assetType = 'website-theme';
    trackCommandUsage('create', { assetType }, derivedAccountId);
    const asset = assets[assetType];
    const argsToPass = {
        commandArgs: args,
        assetType,
        dest,
    };
    if (!argsToPass.dest) {
        argsToPass.dest = resolveLocalPath(asset.dest(argsToPass));
    }
    try {
        await fs.ensureDir(argsToPass.dest);
    }
    catch (e) {
        uiLogger.error(commands.cms.subcommands.theme.subcommands.create.errors.unusablePath(argsToPass.dest));
        logError(e);
        return;
    }
    try {
        await asset.execute(argsToPass);
    }
    catch (e) {
        logError(e);
        process.exit(EXIT_CODES.ERROR);
    }
}
function themeCreateBuilder(yargs) {
    yargs.positional('dest', {
        describe: commands.cms.subcommands.theme.subcommands.create.positionals.dest,
        type: 'string',
    });
    return yargs;
}
const builder = makeYargsBuilder(themeCreateBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
});
const themeCreateCommand = {
    command,
    describe,
    builder,
    handler,
};
export default themeCreateCommand;
