// @hubspot/cli/commands/cms/theme/marketplace-validate.js
import SpinniesManager from '../../../lib/ui/SpinniesManager.js';
import { trackCommandUsage } from '../../../lib/usageTracking.js';
import { kickOffValidation, pollForValidationFinish, fetchValidationResults, processValidationErrors, displayValidationResults, } from '../../../lib/marketplaceValidate.js';
import { commands } from '../../../lang/en.js';
import { makeYargsBuilder } from '../../../lib/yargsUtils.js';
const command = 'marketplace-validate <path>';
const describe = commands.cms.subcommands.theme.subcommands.marketplaceValidate.describe;
async function handler(args) {
    const { path, derivedAccountId } = args;
    trackCommandUsage('validate', {}, derivedAccountId);
    SpinniesManager.init();
    SpinniesManager.add('marketplaceValidation', {
        text: commands.cms.subcommands.theme.subcommands.marketplaceValidate.logs.validatingTheme(path),
    });
    const assetType = 'THEME';
    const validationId = await kickOffValidation(derivedAccountId, assetType, path);
    await pollForValidationFinish(derivedAccountId, validationId);
    SpinniesManager.remove('marketplaceValidation');
    const validationResults = await fetchValidationResults(derivedAccountId, validationId);
    processValidationErrors(commands.cms.subcommands.theme.subcommands.marketplaceValidate.errors
        .invalidPath, validationResults);
    displayValidationResults(commands.cms.subcommands.theme.subcommands.marketplaceValidate.results, validationResults);
    process.exit();
}
function themeValidateBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.theme.subcommands.marketplaceValidate.positionals
            .path.describe,
        type: 'string',
        required: true,
    });
    return yargs;
}
const builder = makeYargsBuilder(themeValidateBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const themeValidateCommand = {
    command,
    describe,
    handler,
    builder,
};
export default themeValidateCommand;
