// @hubspot/cli/commands/theme/generate-selectors.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../lib/ui/index.js';
import generateSelectorsCommand from '../cms/theme/generate-selectors.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
import { commands } from '../../lang/en.js';
const command = 'generate-selectors <path>';
const describe = uiDeprecatedTag(generateSelectorsCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms theme generate-selectors');
    await generateSelectorsCommand.handler(args);
}
function deprecatedThemeSelectorBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.theme.subcommands.generateSelectors.positionals
            .path,
        type: 'string',
        required: true,
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(generateSelectorsCommand.describe, 'hs cms theme generate-selectors');
const builder = makeYargsBuilder(deprecatedThemeSelectorBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
});
const deprecatedGenerateSelectorsCommand = {
    ...generateSelectorsCommand,
    describe,
    handler,
    builder,
};
export default deprecatedGenerateSelectorsCommand;
