// @hubspot/cli/commands/theme/marketplace-validate.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../lib/ui/index.js';
import marketplaceValidateCommand from '../cms/theme/marketplace-validate.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
import { commands } from '../../lang/en.js';
const command = 'marketplace-validate <path>';
const describe = uiDeprecatedTag(marketplaceValidateCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms theme marketplace-validate');
    await marketplaceValidateCommand.handler(args);
}
function deprecatedThemeValidateBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.theme.subcommands.marketplaceValidate.positionals
            .path.describe,
        type: 'string',
        required: true,
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(marketplaceValidateCommand.describe, 'hs cms theme marketplace-validate');
const builder = makeYargsBuilder(deprecatedThemeValidateBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedMarketplaceValidateCommand = {
    ...marketplaceValidateCommand,
    describe,
    handler,
    builder,
};
export default deprecatedMarketplaceValidateCommand;
