// @hubspot/cli/commands/theme/preview.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../lib/ui/index.js';
import previewCommand from '../cms/theme/preview.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
import { commands } from '../../lang/en.js';
const command = 'preview';
const describe = uiDeprecatedTag(previewCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms theme preview');
    await previewCommand.handler(args);
}
function deprecatedThemePreviewBuilder(yargs) {
    yargs
        .option('src', {
        describe: commands.cms.subcommands.theme.subcommands.preview.positionals.src,
        type: 'string',
        requiresArg: true,
    })
        .option('dest', {
        describe: commands.cms.subcommands.theme.subcommands.preview.positionals.dest,
        type: 'string',
        requiresArg: true,
    })
        .option('notify', {
        alias: 'n',
        describe: commands.cms.subcommands.theme.subcommands.preview.options.notify,
        type: 'string',
        requiresArg: true,
    })
        .option('no-ssl', {
        describe: commands.cms.subcommands.theme.subcommands.preview.options.noSsl,
        type: 'boolean',
    })
        .option('port', {
        describe: commands.cms.subcommands.theme.subcommands.preview.options.port,
        type: 'number',
    })
        .option('resetSession', {
        hidden: true,
        type: 'boolean',
    })
        .option('generateFieldsTypes', {
        hidden: true,
        type: 'boolean',
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(previewCommand.describe, 'hs cms theme preview');
const builder = makeYargsBuilder(deprecatedThemePreviewBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
});
const deprecatedPreviewCommand = {
    ...previewCommand,
    describe,
    handler,
    builder,
};
export default deprecatedPreviewCommand;
