// @hubspot/cli/commands/lint.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../lib/ui/index.js';
import cmsLintCommand from './cms/lint.js';
import { makeYargsBuilder } from '../lib/yargsUtils.js';
import { commands } from '../lang/en.js';
const command = 'lint <path>';
const describe = uiDeprecatedTag(cmsLintCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms lint');
    await cmsLintCommand.handler(args);
}
function deprecatedCmsLintBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.lint.positionals.path.describe,
        required: true,
        type: 'string',
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(cmsLintCommand.describe, 'hs cms lint');
const builder = makeYargsBuilder(deprecatedCmsLintBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
});
const deprecatedCmsLintCommand = {
    ...cmsLintCommand,
    describe,
    handler,
    builder,
};
export default deprecatedCmsLintCommand;
