// @hubspot/cli/commands/list.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../lib/ui/index.js';
import cmsListCommand from './cms/list.js';
import { makeYargsBuilder } from '../lib/yargsUtils.js';
import { commands } from '../lang/en.js';
const command = 'list [path]';
const describe = uiDeprecatedTag(cmsListCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms list');
    await cmsListCommand.handler(args);
}
function deprecatedCmsListBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.list.positionals.path.describe,
        type: 'string',
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(cmsListCommand.describe, 'hs cms list');
const builder = makeYargsBuilder(deprecatedCmsListBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedCmsListCommand = {
    ...cmsListCommand,
    describe,
    handler,
    builder,
};
export default deprecatedCmsListCommand;
