// @hubspot/cli/commands/logs.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../lib/ui/index.js';
import functionLogsCommand from './cms/function/logs.js';
import { makeYargsBuilder } from '../lib/yargsUtils.js';
import { commands } from '../lang/en.js';
const command = 'logs [endpoint]';
const describe = uiDeprecatedTag(functionLogsCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms function logs');
    await functionLogsCommand.handler(args);
}
function deprecatedLogsBuilder(yargs) {
    yargs.positional('endpoint', {
        describe: commands.cms.subcommands.function.subcommands.logs.positionals.endpoint
            .describe,
        type: 'string',
    });
    yargs
        .options({
        latest: {
            alias: 'l',
            describe: commands.cms.subcommands.function.subcommands.logs.options.latest
                .describe,
            type: 'boolean',
        },
        compact: {
            describe: commands.cms.subcommands.function.subcommands.logs.options.compact
                .describe,
            type: 'boolean',
        },
        follow: {
            alias: ['f'],
            describe: commands.cms.subcommands.function.subcommands.logs.options.follow
                .describe,
            type: 'boolean',
        },
        limit: {
            describe: commands.cms.subcommands.function.subcommands.logs.options.limit
                .describe,
            type: 'number',
        },
    })
        .conflicts('follow', 'limit');
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(functionLogsCommand.describe, 'hs cms function logs');
const builder = makeYargsBuilder(deprecatedLogsBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedLogsCommand = {
    ...functionLogsCommand,
    describe,
    handler,
    builder,
};
export default deprecatedLogsCommand;
