// @hubspot/cli/commands/mv.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../lib/ui/index.js';
import cmsMvCommand from './cms/mv.js';
import { makeYargsBuilder } from '../lib/yargsUtils.js';
import { commands } from '../lang/en.js';
const command = 'mv <srcPath> <destPath>';
const describe = uiDeprecatedTag(cmsMvCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms mv');
    await cmsMvCommand.handler(args);
}
function deprecatedCmsMvBuilder(yargs) {
    yargs.positional('srcPath', {
        describe: commands.cms.subcommands.mv.positionals.srcPath.describe,
        type: 'string',
    });
    yargs.positional('destPath', {
        describe: commands.cms.subcommands.mv.positionals.destPath.describe,
        type: 'string',
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(cmsMvCommand.describe, 'hs cms mv');
const builder = makeYargsBuilder(deprecatedCmsMvBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedCmsMvCommand = {
    ...cmsMvCommand,
    describe,
    handler,
    builder,
};
export default deprecatedCmsMvCommand;
