// @hubspot/cli/commands/remove.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../lib/ui/index.js';
import cmsDeleteCommand from './cms/delete.js';
import { makeYargsBuilder } from '../lib/yargsUtils.js';
import { commands } from '../lang/en.js';
const command = 'remove <path>';
const describe = uiDeprecatedTag(cmsDeleteCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms delete');
    await cmsDeleteCommand.handler(args);
}
function deprecatedCmsRemoveBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.delete.positionals.path,
        type: 'string',
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(cmsDeleteCommand.describe, 'hs cms delete');
const builder = makeYargsBuilder(deprecatedCmsRemoveBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedCmsRemoveCommand = {
    command,
    describe,
    handler,
    builder,
};
export default deprecatedCmsRemoveCommand;
