// @hubspot/cli/commands/upload.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../lib/ui/index.js';
import cmsUploadCommand from './cms/upload.js';
import { makeYargsBuilder } from '../lib/yargsUtils.js';
import { commands } from '../lang/en.js';
const command = 'upload [src] [dest]';
const describe = uiDeprecatedTag(cmsUploadCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms upload');
    await cmsUploadCommand.handler(args);
}
function deprecatedCmsUploadBuilder(yargs) {
    yargs.positional('src', {
        describe: commands.cms.subcommands.upload.positionals.src,
        type: 'string',
    });
    yargs.positional('dest', {
        describe: commands.cms.subcommands.upload.positionals.dest,
        type: 'string',
    });
    yargs.option('fieldOptions', {
        describe: commands.cms.subcommands.upload.options.options,
        type: 'array',
        default: [''],
        hidden: true,
    });
    yargs.option('saveOutput', {
        describe: commands.cms.subcommands.upload.options.saveOutput,
        type: 'boolean',
        default: false,
    });
    yargs.option('convertFields', {
        describe: commands.cms.subcommands.upload.options.convertFields,
        type: 'boolean',
        default: false,
    });
    yargs.option('clean', {
        describe: commands.cms.subcommands.upload.options.clean,
        type: 'boolean',
        default: false,
    });
    yargs.option('force', {
        describe: commands.cms.subcommands.upload.options.force,
        type: 'boolean',
        default: false,
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(cmsUploadCommand.describe, 'hs cms upload');
const builder = makeYargsBuilder(deprecatedCmsUploadBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
    useCmsPublishModeOptions: true,
});
const deprecatedCmsUploadCommand = {
    ...cmsUploadCommand,
    describe,
    handler,
    builder,
};
export default deprecatedCmsUploadCommand;
