// @hubspot/app-functions-dev-server/dist/api/privateAppUserToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updatePrivateAppUserToken = exports.createPrivateAppUserToken = exports.fetchPrivateAppUserToken = void 0;
const http_1 = require("@hubspot/local-dev-lib/http");
const LOCALDEVAUTH_API_PRIVATE_APP_USER_TOKEN_PATH = 'localdevauth/v1/private-app/user-token';
async function fetchPrivateAppUserToken({ accountId, appId, }) {
    return await http_1.http.get(accountId, {
        url: `${LOCALDEVAUTH_API_PRIVATE_APP_USER_TOKEN_PATH}/${appId}`,
    });
}
exports.fetchPrivateAppUserToken = fetchPrivateAppUserToken;
async function createPrivateAppUserToken({ accountId, appId, scopeGroups, expiresAt, privateAppToken, }) {
    return await http_1.http.post(accountId, {
        url: `${LOCALDEVAUTH_API_PRIVATE_APP_USER_TOKEN_PATH}/${appId}`,
        data: {
            ...(scopeGroups && { scopeGroups }),
            ...(expiresAt && { expiresAt }),
            ...(privateAppToken && { privateAppToken }),
        },
    });
}
exports.createPrivateAppUserToken = createPrivateAppUserToken;
async function updatePrivateAppUserToken({ accountId, appId, userTokenKey, scopeGroups, expiresAt, privateAppToken, }) {
    return await http_1.http.put(accountId, {
        url: `${LOCALDEVAUTH_API_PRIVATE_APP_USER_TOKEN_PATH}/${appId}`,
        data: {
            privateAppUserTokenKey: userTokenKey,
            ...(scopeGroups && { scopeGroups }),
            ...(expiresAt && { expiresAt }),
            ...(privateAppToken && { privateAppToken }),
        },
    });
}
exports.updatePrivateAppUserToken = updatePrivateAppUserToken;
