// @hubspot/cli/mcp-server/tools/project/GetBuildLogsTool.js
import { Tool } from '../../types.js';
import { z } from 'zod';
import { trackToolUsage } from '../../utils/toolUsageTracking.js';
import { formatTextContents } from '../../utils/content.js';
import { isHubSpotHttpError } from '@hubspot/local-dev-lib/errors/index';
import { http } from '@hubspot/local-dev-lib/http';
import { getProjectConfig, validateProjectConfig, } from '../../../lib/projects/config.js';
import { absoluteCurrentWorkingDirectory, absoluteProjectPath, } from './constants.js';
import { getConfigDefaultAccountIfExists } from '@hubspot/local-dev-lib/config';
import { setupHubSpotConfig } from '../../utils/config.js';
const TOOL_NAME = 'get-build-logs';
const PROJECTS_LOGS_API_PATH = 'dfs/logging/v1';
const inputSchema = {
    absoluteProjectPath,
    absoluteCurrentWorkingDirectory,
    buildId: z
        .number()
        .describe('Build ID to fetch logs for. Use get-build-status to find recent build IDs.'),
    logLevel: z
        .enum(['ERROR', 'WARN', 'INFO', 'ALL'])
        .optional()
        .default('ALL')
        .describe('Filter logs by level. ERROR: Show only errors, WARN: Show only warnings, INFO: Show only info, ALL: Show all logs.'),
};
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const inputSchemaZodObject = z.object({ ...inputSchema });
function flattenLogs(response) {
    const allLogs = [];
    response.logs.forEach(substep => {
        substep.logs.forEach(log => {
            allLogs.push(log);
        });
    });
    return allLogs.sort((a, b) => a.timestamp - b.timestamp);
}
function filterLogsByLevel(logs, logLevel) {
    if (logLevel === 'ALL') {
        return logs;
    }
    return logs.filter(log => log.logLevel === logLevel);
}
function formatLogs(logs) {
    if (logs.length === 0) {
        return 'No logs found.';
    }
    return logs
        .map(log => {
        const timestamp = new Date(log.timestamp).toLocaleString('en-US', {
            month: 'short',
            day: 'numeric',
            year: 'numeric',
            hour: 'numeric',
            minute: 'numeric',
            second: 'numeric',
            timeZoneName: 'short',
        });
        const component = log.pipelineSubstepName || `Step ${log.pipelineStepId}`;
        return `[${log.logLevel}][${component}] ${timestamp} ${log.message}`;
    })
        .join('\n');
}
export class GetBuildLogsTool extends Tool {
    constructor(mcpServer) {
        super(mcpServer);
    }
    async handler({ absoluteProjectPath, absoluteCurrentWorkingDirectory, buildId, logLevel, }) {
        setupHubSpotConfig(absoluteCurrentWorkingDirectory);
        await trackToolUsage(TOOL_NAME);
        try {
            const accountId = getConfigDefaultAccountIfExists()?.accountId;
            if (!accountId) {
                return formatTextContents(absoluteCurrentWorkingDirectory, 'No account ID found. Please run `hs account auth` to configure an account, or set a default account with `hs account use <account>`');
            }
            const { projectConfig, projectDir } = await getProjectConfig(absoluteProjectPath);
            validateProjectConfig(projectConfig, projectDir);
            const projectName = projectConfig.name;
            const response = await http.get(accountId, {
                url: `${PROJECTS_LOGS_API_PATH}/logs/projects/${encodeURIComponent(projectName)}/builds/${buildId}`,
            });
            const buildLogsResponse = response.data;
            if (!buildLogsResponse.logs || buildLogsResponse.logs.length === 0) {
                return formatTextContents(absoluteCurrentWorkingDirectory, `No logs found for build #${buildId} in project '${projectName}'.`);
            }
            const allLogs = flattenLogs(buildLogsResponse);
            if (allLogs.length === 0) {
                return formatTextContents(absoluteCurrentWorkingDirectory, `No logs found for build #${buildId} in project '${projectName}'.`);
            }
            const filteredLogs = filterLogsByLevel(allLogs, logLevel);
            let output;
            if (filteredLogs.length === 0) {
                // No logs match filter, show all logs instead
                output = `No ${logLevel} level logs found for build #${buildId} in '${projectName}'.\nShowing all logs instead:\n\n${formatLogs(allLogs)}`;
            }
            else {
                output = `Logs for build #${buildId} in '${projectName}' (${logLevel} level):\n\n${formatLogs(filteredLogs)}`;
            }
            return formatTextContents(absoluteCurrentWorkingDirectory, output);
        }
        catch (error) {
            let errorMessage;
            if (isHubSpotHttpError(error)) {
                errorMessage = error.toString();
            }
            else if (error instanceof Error) {
                errorMessage = error.message;
            }
            else {
                errorMessage = String(error);
            }
            return formatTextContents(absoluteCurrentWorkingDirectory, errorMessage);
        }
    }
    register() {
        return this.mcpServer.registerTool(TOOL_NAME, {
            title: 'Get HubSpot Project Build Logs',
            description: 'Retrieves build logs for a specific HubSpot project build. Use this to debug build failures by viewing the full build pipeline output. This tool is for more comprehensive troubleshootings or addressing build WARNINGs, build errors should be troubleshooted with get-build-status tool first. Logs can be filtered by level (ERROR, WARN, INFO, or ALL). Use `hs project list-builds`  first to identify the build ID and error messages.',
            inputSchema,
            annotations: {
                readOnlyHint: true,
                openWorldHint: true,
                idempotentHint: true,
            },
        }, this.handler);
    }
}
