// @hubspot/cli/mcp-server/tools/project/ValidateProjectTool.js
import { Tool } from '../../types.js';
import { z } from 'zod';
import { absoluteCurrentWorkingDirectory, absoluteProjectPath, } from './constants.js';
import { runCommandInDir } from '../../utils/project.js';
import { formatTextContents } from '../../utils/content.js';
import { trackToolUsage } from '../../utils/toolUsageTracking.js';
import { setupHubSpotConfig } from '../../utils/config.js';
const inputSchema = {
    absoluteProjectPath,
    absoluteCurrentWorkingDirectory,
};
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const inputSchemaZodObject = z.object({ ...inputSchema });
const toolName = 'validate-project';
export class ValidateProjectTool extends Tool {
    constructor(mcpServer) {
        super(mcpServer);
    }
    async handler({ absoluteProjectPath, absoluteCurrentWorkingDirectory, }) {
        setupHubSpotConfig(absoluteCurrentWorkingDirectory);
        await trackToolUsage(toolName);
        try {
            const { stdout, stderr } = await runCommandInDir(absoluteProjectPath, 'hs project validate');
            return formatTextContents(stdout, stderr);
        }
        catch (error) {
            return formatTextContents(error instanceof Error ? error.message : `${error}`);
        }
    }
    register() {
        return this.mcpServer.registerTool(toolName, {
            title: 'Validate HubSpot Project',
            description: 'Validates the HubSpot project and its configuration files.  This tool does not need to be ran before uploading the project',
            inputSchema,
            annotations: {
                readOnlyHint: true,
                openWorldHint: false,
            },
        }, this.handler);
    }
}
