// @hubspot/cli/commands/filemanager/fetch.js
import { downloadFileOrFolder } from '@hubspot/local-dev-lib/fileManager';
import { uiLogger } from '../../lib/ui/logger.js';
import { resolveLocalPath } from '../../lib/filesystem.js';
import { trackCommandUsage } from '../../lib/usageTracking.js';
import { commands } from '../../lang/en.js';
import { EXIT_CODES } from '../../lib/enums/exitCodes.js';
import { logError } from '../../lib/errorHandlers/index.js';
import { addOverwriteOptions } from '../../lib/commonOpts.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
const command = 'fetch <src> [dest]';
const describe = commands.filemanager.subcommands.fetch.describe;
async function handler(args) {
    const { src, includeArchived, derivedAccountId, overwrite } = args;
    if (typeof src !== 'string') {
        uiLogger.error(commands.filemanager.subcommands.fetch.errors.sourceRequired);
        process.exit(EXIT_CODES.ERROR);
    }
    const dest = resolveLocalPath(args.dest);
    trackCommandUsage('filemanager-fetch', {}, derivedAccountId);
    try {
        // Fetch and write file/folder.
        await downloadFileOrFolder(derivedAccountId, src, dest, overwrite, includeArchived || false);
    }
    catch (err) {
        logError(err);
        process.exit(EXIT_CODES.ERROR);
    }
}
function fileManagerFetchBuilder(yargs) {
    addOverwriteOptions(yargs);
    yargs.positional('src', {
        describe: commands.filemanager.subcommands.fetch.positionals.src.describe,
        type: 'string',
    });
    yargs.positional('dest', {
        describe: commands.filemanager.subcommands.fetch.positionals.dest.describe,
        type: 'string',
    });
    yargs.option('include-archived', {
        alias: ['i'],
        describe: commands.filemanager.subcommands.fetch.options.includeArchived.describe,
        type: 'boolean',
    });
    return yargs;
}
const builder = makeYargsBuilder(fileManagerFetchBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const fileManagerFetchCommand = {
    command,
    describe,
    handler,
    builder,
};
export default fileManagerFetchCommand;
