// @hubspot/app-functions-dev-server/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.localProxyErrorMappings = exports.axiosErrorMappings = exports.validationError = exports.defaultServerError = exports.PLATFORM_VERSION = exports.SECRETS_IN_CONTEXT = exports.PRIVATE_APP_ACCESS_TOKEN = exports.EXECUTION_TIMEOUT_MS = exports.DEV_SERVER_DEFAULT_PORT = exports.APP_FUNCTIONS_MANIFEST_FILENAME = exports.APP_FUNCTIONS_DIRNAME = void 0;
const axios_1 = require("axios");
exports.APP_FUNCTIONS_DIRNAME = 'app.functions';
exports.APP_FUNCTIONS_MANIFEST_FILENAME = 'serverless.json';
exports.DEV_SERVER_DEFAULT_PORT = 6789;
exports.EXECUTION_TIMEOUT_MS = 15000;
exports.PRIVATE_APP_ACCESS_TOKEN = 'PRIVATE_APP_ACCESS_TOKEN';
exports.SECRETS_IN_CONTEXT = [
    exports.PRIVATE_APP_ACCESS_TOKEN,
    'HS_ENVIRONMENT',
    'HS_HUBLET',
    'HS_SERVERLESS_FUNCTION_ID',
];
exports.PLATFORM_VERSION = {
    V20231: '2023.1',
    V20232: '2023.2',
    V20251: '2025.1',
    V20252: '2025.2',
    UNSTABLE: 'unstable',
};
exports.defaultServerError = {
    status: axios_1.HttpStatusCode.InternalServerError,
    message: 'Failure observed in UI Extensions Proxy Server. Contact Hubspot Support if the problem persists',
    category: 'INTERNAL_SERVER_ERROR',
};
exports.validationError = {
    category: 'VALIDATION_ERROR',
    status: axios_1.HttpStatusCode.BadRequest,
};
// Axios errors types: https://github.com/axios/axios/blob/v1.x/README.md#error-types
exports.axiosErrorMappings = {
    [axios_1.AxiosError.ERR_FR_TOO_MANY_REDIRECTS]: {
        ...exports.defaultServerError,
        message: 'Too many redirects',
    },
    [axios_1.AxiosError.ERR_NETWORK]: {
        ...exports.defaultServerError,
        message: 'Network error',
    },
    [axios_1.AxiosError.ERR_BAD_RESPONSE]: {
        ...exports.validationError,
        message: 'Invalid JSON Response Body',
    },
    [axios_1.AxiosError.ERR_BAD_REQUEST]: {
        ...exports.validationError,
        message: 'Invalid JSON Request Body',
    },
    [axios_1.AxiosError.ERR_INVALID_URL]: {
        ...exports.validationError,
        message: 'URL is malformed',
    },
    [axios_1.AxiosError.ETIMEDOUT]: {
        status: axios_1.HttpStatusCode.GatewayTimeout,
        category: 'GATEWAY_TIMEOUT',
        message: 'Gateway took too long to respond',
    },
};
exports.localProxyErrorMappings = {
    BAD_GATEWAY: {
        status: 200,
        category: 'BAD_GATEWAY',
        message: 'Error returned from Request Uri',
    },
};
