// @hubspot/app-functions-dev-server/dist/signing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSignatureHeaders = exports.formatForSigning = exports.sign = void 0;
const api_client_1 = require("@hubspot/api-client");
// Only exported for testing
exports.sign = {
    v2: ({ method, ...options }) => api_client_1.Signature.getSignature(method, 'v2', options),
    v3: ({ method, ...options }) => api_client_1.Signature.getSignature(method, 'v3', options),
};
const formatForSigning = (body) => {
    if (body === null) {
        return '';
    }
    if (typeof body === 'number' &&
        (!Number.isFinite(body) || Number.isNaN(body))) {
        return '';
    }
    // Anything that stringifies to undefined will be an empty string
    return JSON.stringify(body) || '';
};
exports.formatForSigning = formatForSigning;
// This is only used for signing local dev proxy requests
const getSignatureHeaders = ({ method, url, requestBody, }) => {
    const clientSecret = process.env.CLIENT_SECRET;
    if (!clientSecret) {
        return;
    }
    const timestamp = new Date().getTime();
    const signingProps = {
        method,
        url,
        requestBody: (0, exports.formatForSigning)(requestBody),
        clientSecret,
        /**
         * I'm not sure why the signing function requires this
         *
         * https://github.com/HubSpot/hubspot-api-nodejs/blob/bfaa106af7199adbce5f766b1eb35a73f3c72499/src/utils/ISignatureOptions.ts
         *
         * But it doesn't use it for signing
         *
         * https://github.com/HubSpot/hubspot-api-nodejs/blob/bfaa106af7199adbce5f766b1eb35a73f3c72499/src/utils/signature.ts#L19-L32
         *
         */
        signature: '',
    };
    return {
        'X-HubSpot-Signature': exports.sign.v2(signingProps),
        'X-HubSpot-Signature-Version': 'v2',
        'X-HubSpot-Request-Timestamp': timestamp,
        'X-HubSpot-Signature-v3': exports.sign.v3({
            ...signingProps,
            timestamp,
        }),
    };
};
exports.getSignatureHeaders = getSignatureHeaders;
