// @hubspot/cli/commands/customObject/schema/create.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../../lib/ui/index.js';
import createSchemaCommand from '../createSchema.js';
import { makeYargsBuilder } from '../../../lib/yargsUtils.js';
import { commands } from '../../../lang/en.js';
const command = 'create';
const describe = uiDeprecatedTag(createSchemaCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs custom-object create-schema');
    await createSchemaCommand.handler(args);
}
function deprecatedCreateSchemaBuilder(yargs) {
    yargs.option('path', {
        describe: commands.customObject.subcommands.createSchema.options.definition
            .describe,
        type: 'string',
        required: true,
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(createSchemaCommand.describe, 'hs custom-object create-schema');
const builder = makeYargsBuilder(deprecatedCreateSchemaBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
    useTestingOptions: true,
});
const deprecatedCreateSchemaCommand = {
    command,
    describe,
    handler,
    builder,
};
export default deprecatedCreateSchemaCommand;
