// @hubspot/cli/commands/customObject/schema/delete.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../../lib/ui/index.js';
import deleteSchemaCommand from '../deleteSchema.js';
import { makeYargsBuilder } from '../../../lib/yargsUtils.js';
import { commands } from '../../../lang/en.js';
const command = 'delete <name>';
const describe = uiDeprecatedTag(deleteSchemaCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs custom-object delete-schema');
    await deleteSchemaCommand.handler(args);
}
function deprecatedDeleteSchemaBuilder(yargs) {
    yargs
        .positional('name', {
        describe: commands.customObject.subcommands.deleteSchema.positionals.name
            .describe,
        type: 'string',
    })
        .option('force', {
        describe: commands.customObject.subcommands.deleteSchema.options.force.describe,
        type: 'boolean',
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(deleteSchemaCommand.describe, 'hs custom-object delete-schema');
const builder = makeYargsBuilder(deprecatedDeleteSchemaBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedDeleteSchemaCommand = {
    command,
    describe,
    handler,
    builder,
};
export default deprecatedDeleteSchemaCommand;
