// @hubspot/cli/commands/customObject/schema/fetch-all.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../../lib/ui/index.js';
import fetchAllSchemasCommand from '../fetchAllSchemas.js';
import { makeYargsBuilder } from '../../../lib/yargsUtils.js';
import { commands } from '../../../lang/en.js';
const command = 'fetch-all [dest]';
const describe = uiDeprecatedTag(fetchAllSchemasCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs custom-object fetch-all-schemas');
    await fetchAllSchemasCommand.handler(args);
}
function deprecatedFetchAllSchemasBuilder(yargs) {
    yargs.positional('dest', {
        describe: commands.customObject.subcommands.fetchAllSchemas.positionals.dest
            .describe,
        type: 'string',
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(fetchAllSchemasCommand.describe, 'hs custom-object fetch-all-schemas');
const builder = makeYargsBuilder(deprecatedFetchAllSchemasBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedFetchAllSchemasCommand = {
    command,
    describe,
    handler,
    builder,
};
export default deprecatedFetchAllSchemasCommand;
