// @hubspot/cli/commands/customObject/schema/fetch.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../../lib/ui/index.js';
import fetchSchemaCommand from '../fetchSchema.js';
import { makeYargsBuilder } from '../../../lib/yargsUtils.js';
import { commands } from '../../../lang/en.js';
const command = 'fetch <name> [dest]';
const describe = uiDeprecatedTag(fetchSchemaCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs custom-object fetch-schema');
    await fetchSchemaCommand.handler(args);
}
function deprecatedFetchSchemaBuilder(yargs) {
    yargs
        .positional('name', {
        describe: commands.customObject.subcommands.fetchSchema.positionals.name.describe,
        type: 'string',
    })
        .positional('dest', {
        describe: commands.customObject.subcommands.fetchSchema.positionals.dest.describe,
        type: 'string',
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(fetchSchemaCommand.describe, 'hs custom-object fetch-schema');
const builder = makeYargsBuilder(deprecatedFetchSchemaBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedFetchSchemaCommand = {
    command,
    describe,
    handler,
    builder,
};
export default deprecatedFetchSchemaCommand;
