// @hubspot/cli/commands/customObject/schema/update.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../../lib/ui/index.js';
import updateSchemaCommand from '../updateSchema.js';
import { makeYargsBuilder } from '../../../lib/yargsUtils.js';
import { commands } from '../../../lang/en.js';
const command = 'update <name>';
const describe = uiDeprecatedTag(updateSchemaCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs custom-object update-schema');
    await updateSchemaCommand.handler(args);
}
function deprecatedUpdateSchemaBuilder(yargs) {
    yargs
        .positional('name', {
        describe: commands.customObject.subcommands.updateSchema.positionals.name
            .describe,
        type: 'string',
    })
        .option('path', {
        describe: commands.customObject.subcommands.updateSchema.options.path.describe,
        type: 'string',
        required: true,
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(updateSchemaCommand.describe, 'hs custom-object update-schema');
const builder = makeYargsBuilder(deprecatedUpdateSchemaBuilder, command, verboseDescribe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
    useTestingOptions: true,
});
const deprecatedUpdateSchemaCommand = {
    command,
    describe,
    handler,
    builder,
};
export default deprecatedUpdateSchemaCommand;
