// @hubspot/cli/commands/module/marketplace-validate.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../lib/ui/index.js';
import marketplaceValidateCommand from '../cms/module/marketplace-validate.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
import { commands } from '../../lang/en.js';
const command = 'marketplace-validate [src]';
const describe = uiDeprecatedTag(marketplaceValidateCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms module marketplace-validate');
    await marketplaceValidateCommand.handler(args);
}
function deprecatedMarketplaceValidateBuilder(yargs) {
    yargs.positional('src', {
        describe: commands.cms.subcommands.module.subcommands.marketplaceValidate
            .positionals.src,
        type: 'string',
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(marketplaceValidateCommand.describe, 'hs cms module marketplace-validate');
const builder = makeYargsBuilder(deprecatedMarketplaceValidateBuilder, command, verboseDescribe, {
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedMarketplaceValidateCommand = {
    ...marketplaceValidateCommand,
    describe,
    handler,
    builder,
};
export default deprecatedMarketplaceValidateCommand;
