// @hubspot/cli/lib/projects/config.js
import fs from 'fs-extra';
import path from 'path';
import findup from 'findup-sync';
import { getAbsoluteFilePath, getCwd } from '@hubspot/local-dev-lib/path';
import { PROJECT_CONFIG_FILE } from '../constants.js';
import { lib } from '../../lang/en.js';
import { uiLogger } from '../ui/logger.js';
import ProjectValidationError from '../errors/ProjectValidationError.js';
export function writeProjectConfig(configPath, config) {
    try {
        fs.ensureFileSync(configPath);
        fs.writeFileSync(configPath, JSON.stringify(config, null, 2));
        uiLogger.debug(`Wrote project config at ${configPath}`);
    }
    catch (e) {
        uiLogger.debug(e);
        return false;
    }
    return true;
}
export function getIsInProject(dir) {
    const configPath = getProjectConfigPath(dir);
    return !!configPath;
}
function getProjectConfigPath(dir) {
    const projectDir = dir ? getAbsoluteFilePath(dir) : getCwd();
    const configPath = findup(PROJECT_CONFIG_FILE, {
        cwd: projectDir,
        nocase: true,
    });
    return configPath;
}
export async function getProjectConfig(dir) {
    const configPath = getProjectConfigPath(dir);
    if (!configPath) {
        return { projectConfig: null, projectDir: null };
    }
    try {
        const config = fs.readFileSync(configPath);
        const projectConfig = JSON.parse(config.toString());
        return {
            projectDir: path.dirname(configPath),
            projectConfig,
        };
    }
    catch (e) {
        uiLogger.error(lib.projects.getProjectConfig.error);
        return { projectConfig: null, projectDir: null };
    }
}
export function validateProjectConfig(projectConfig, projectDir) {
    if (!projectConfig || !projectDir) {
        throw new ProjectValidationError(lib.projects.validateProjectConfig.configNotFound);
    }
    if (!projectConfig.name || !projectConfig.srcDir) {
        throw new ProjectValidationError(lib.projects.validateProjectConfig.configMissingFields);
    }
    const resolvedPath = path.resolve(projectDir, projectConfig.srcDir);
    if (!resolvedPath.startsWith(projectDir)) {
        const projectConfigFile = path.relative('.', path.join(projectDir, PROJECT_CONFIG_FILE));
        throw new ProjectValidationError(lib.projects.validateProjectConfig.srcOutsideProjectDir(projectConfigFile, projectConfig.srcDir));
    }
    if (!fs.existsSync(resolvedPath)) {
        throw new ProjectValidationError(lib.projects.validateProjectConfig.srcDirNotFound(projectConfig.srcDir, projectDir));
    }
}
