// @hubspot/cli/lib/projects/urls.js
import { getHubSpotWebsiteOrigin } from '@hubspot/local-dev-lib/urls';
import { getConfigAccountEnvironment } from '@hubspot/local-dev-lib/config';
function getBaseUrl(accountId) {
    return getHubSpotWebsiteOrigin(getConfigAccountEnvironment(accountId));
}
function getProjectHomeUrl(accountId) {
    return `${getBaseUrl(accountId)}/developer-projects/${accountId}`;
}
export function getProjectComponentDistributionUrl(projectName, componentName, accountId) {
    const baseUrl = getHubSpotWebsiteOrigin(getConfigAccountEnvironment(accountId));
    return `${baseUrl}/developer-projects/${accountId}/project/${projectName}/component/${componentName}/distribution`;
}
export function getDeveloperOverviewUrl(accountId) {
    const baseUrl = getHubSpotWebsiteOrigin(getConfigAccountEnvironment(accountId));
    return `${baseUrl}/developer-overview/${accountId}`;
}
export function getProjectDetailUrl(projectName, accountId) {
    if (!projectName)
        return;
    return `${getProjectHomeUrl(accountId)}/project/${projectName}`;
}
export function getProjectSettingsUrl(projectName, accountId) {
    if (!projectName)
        return;
    return `${getProjectDetailUrl(projectName, accountId)}/settings`;
}
export function getProjectActivityUrl(projectName, accountId) {
    return `${getProjectDetailUrl(projectName, accountId)}/activity`;
}
export function getProjectBuildDetailUrl(projectName, buildId, accountId) {
    return `${getProjectActivityUrl(projectName, accountId)}/build/${buildId}`;
}
export function getProjectDeployDetailUrl(projectName, deployId, accountId) {
    return `${getProjectActivityUrl(projectName, accountId)}/deploy/${deployId}`;
}
export function getLocalDevUiUrl(accountId, showWelcomeScreen) {
    return `${getBaseUrl(accountId)}/developer-projects-local-dev/${accountId}${showWelcomeScreen ? '?welcome' : ''}`;
}
export function getAccountHomeUrl(accountId) {
    const baseUrl = getHubSpotWebsiteOrigin(getConfigAccountEnvironment(accountId));
    return `${baseUrl}/home?portalId=${accountId}`;
}
