// @hubspot/cli/lib/projects/localDev/DevServerManager_DEPRECATED.js
import { logger } from '@hubspot/local-dev-lib/logger';
import { DevModeInterface as UIEDevModeInterface } from '@hubspot/ui-extensions-dev-server';
import { startPortManagerServer, stopPortManagerServer, requestPorts, } from '@hubspot/local-dev-lib/portManager';
import { getHubSpotApiOrigin, getHubSpotWebsiteOrigin, } from '@hubspot/local-dev-lib/urls';
import { getConfigAccountById } from '@hubspot/local-dev-lib/config';
import { ComponentTypes, } from '../../../types/Projects.js';
import { lib } from '../../../lang/en.js';
import { uiLogger } from '../../ui/logger.js';
import { logError } from '../../errorHandlers/index.js';
import { EXIT_CODES } from '../../enums/exitCodes.js';
const SERVER_KEYS = {
    privateApp: 'privateApp',
    publicApp: 'publicApp',
};
class DevServerManager_DEPRECATED {
    initialized;
    started;
    componentsByType;
    devServers;
    constructor() {
        this.initialized = false;
        this.started = false;
        this.componentsByType = {};
        this.devServers = {
            [SERVER_KEYS.privateApp]: {
                componentType: ComponentTypes.PrivateApp,
                serverInterface: UIEDevModeInterface,
            },
            [SERVER_KEYS.publicApp]: {
                componentType: ComponentTypes.PublicApp,
                serverInterface: UIEDevModeInterface,
            },
        };
    }
    async iterateDevServers(callback) {
        const serverKeys = Object.keys(this.devServers);
        for (let i = 0; i < serverKeys.length; i++) {
            const serverKey = serverKeys[i];
            const devServer = this.devServers[serverKey];
            const compatibleComponents = this.componentsByType[devServer.componentType] || {};
            if (Object.keys(compatibleComponents).length) {
                await callback(devServer.serverInterface, compatibleComponents);
            }
            else {
                uiLogger.debug(lib.DevServerManager.noCompatibleComponents(serverKey));
            }
        }
    }
    arrangeComponentsByType(components) {
        return components.reduce((acc, component) => {
            if (!acc[component.type]) {
                acc[component.type] = {};
            }
            if ('name' in component.config && component.config.name) {
                acc[component.type][component.config.name] = component;
            }
            return acc;
        }, {});
    }
    async setup({ components, onUploadRequired, accountId, setActiveApp, }) {
        this.componentsByType = this.arrangeComponentsByType(components);
        let env;
        const accountConfig = getConfigAccountById(accountId);
        if (accountConfig) {
            env = accountConfig.env;
        }
        try {
            await startPortManagerServer();
        }
        catch (e) {
            logError(e);
            process.exit(EXIT_CODES.ERROR);
        }
        await this.iterateDevServers(async (serverInterface, compatibleComponents) => {
            if (serverInterface.setup) {
                await serverInterface.setup({
                    components: compatibleComponents,
                    onUploadRequired,
                    logger,
                    urls: {
                        api: getHubSpotApiOrigin(env),
                        web: getHubSpotWebsiteOrigin(env),
                    },
                    setActiveApp,
                });
            }
        });
        this.initialized = true;
    }
    async start({ accountId, projectConfig, }) {
        if (this.initialized) {
            await this.iterateDevServers(async (serverInterface) => {
                if (serverInterface.start) {
                    await serverInterface.start({
                        accountId,
                        projectConfig,
                        requestPorts,
                    });
                }
            });
        }
        else {
            throw new Error(lib.DevServerManager.notInitialized);
        }
        this.started = true;
    }
    async fileChange({ filePath, event, }) {
        if (this.started) {
            this.iterateDevServers(async (serverInterface) => {
                if (serverInterface.fileChange) {
                    await serverInterface.fileChange(filePath, event);
                }
            });
        }
    }
    async cleanup() {
        if (this.started) {
            await this.iterateDevServers(async (serverInterface) => {
                if (serverInterface.cleanup) {
                    await serverInterface.cleanup();
                }
            });
            await stopPortManagerServer();
        }
    }
}
const Manager = new DevServerManager_DEPRECATED();
export default Manager;
