// @hubspot/cli/lib/projects/localDev/DevServerManager.js
import AppDevModeInterface from './AppDevModeInterface.js';
import { lib } from '../../../lang/en.js';
import UIExtensionsDevModeInterface from './UIExtensionsDevModeInterface.js';
class DevServerManager {
    initialized;
    started;
    devServers;
    constructor(options) {
        this.initialized = false;
        this.started = false;
        const AppsDevServer = new AppDevModeInterface({
            localDevState: options.localDevState,
            localDevLogger: options.logger,
        });
        const UIExtensionsDevServer = new UIExtensionsDevModeInterface({
            localDevState: options.localDevState,
        });
        this.devServers = [AppsDevServer, UIExtensionsDevServer];
    }
    async iterateDevServers(callback) {
        await Promise.all(this.devServers.map(devServer => callback(devServer)));
    }
    async setup() {
        for (const devServer of this.devServers) {
            if (devServer.setup) {
                // Run setup functions in order
                await devServer.setup();
            }
        }
        this.initialized = true;
    }
    async start() {
        if (this.initialized) {
            await this.iterateDevServers(async (serverInterface) => {
                if (serverInterface.start) {
                    await serverInterface.start();
                }
            });
        }
        else {
            throw new Error(lib.DevServerManager.notInitialized);
        }
        this.started = true;
    }
    async fileChange({ filePath, event, }) {
        if (this.started) {
            this.iterateDevServers(async (serverInterface) => {
                if (serverInterface.fileChange) {
                    await serverInterface.fileChange(filePath, event);
                }
            });
        }
    }
    async cleanup() {
        if (this.started) {
            await this.iterateDevServers(async (serverInterface) => {
                if (serverInterface.cleanup) {
                    await serverInterface.cleanup();
                }
            });
        }
    }
}
export default DevServerManager;
