// @hubspot/cli/lib/projects/localDev/LocalDevWatcher.js
import path from 'path';
import chokidar from 'chokidar';
import { PROJECT_CONFIG_FILE } from '../../constants.js';
const WATCH_EVENTS = {
    add: 'add',
    change: 'change',
    unlink: 'unlink',
    unlinkDir: 'unlinkDir',
};
class LocalDevWatcher {
    localDevProcess;
    watcher;
    constructor(localDevProcess) {
        this.localDevProcess = localDevProcess;
        this.watcher = null;
    }
    handleWatchEvent(filePath, event, configPaths) {
        if (configPaths.includes(filePath)) {
            return this.localDevProcess.handleConfigFileChange();
        }
        return this.localDevProcess.handleFileChange(filePath, event);
    }
    start() {
        this.watcher = chokidar.watch(this.localDevProcess.projectDir, {
            ignoreInitial: true,
            ignored: ['**/dist'],
        });
        const configPaths = Object.values(this.localDevProcess.projectNodes).map(component => component.localDev.componentConfigPath);
        const projectConfigPath = path.join(this.localDevProcess.projectDir, PROJECT_CONFIG_FILE);
        configPaths.push(projectConfigPath);
        this.watcher.on('add', filePath => {
            this.handleWatchEvent(filePath, WATCH_EVENTS.add, configPaths);
        });
        this.watcher.on('change', filePath => {
            this.handleWatchEvent(filePath, WATCH_EVENTS.change, configPaths);
        });
        this.watcher.on('unlink', filePath => {
            this.handleWatchEvent(filePath, WATCH_EVENTS.unlink, configPaths);
        });
        this.watcher.on('unlinkDir', filePath => {
            this.handleWatchEvent(filePath, WATCH_EVENTS.unlinkDir, configPaths);
        });
    }
    async stop() {
        await this.watcher?.close();
    }
}
export default LocalDevWatcher;
