// @hubspot/cli/lib/projects/localDev/LocalDevState.js
import { LOCAL_DEV_SERVER_MESSAGE_TYPES } from '../../constants.js';
class LocalDevState {
    _targetProjectAccountId;
    _targetTestingAccountId;
    _profile;
    _projectConfig;
    _projectDir;
    _projectData;
    _debug;
    _projectNodes;
    _projectProfileData;
    _projectNodesAtLastDeploy;
    _env;
    _listeners;
    _appData;
    _devServerMessage;
    _uploadWarnings;
    _devServersStarted;
    constructor({ targetProjectAccountId, targetTestingAccountId, projectConfig, projectDir, projectData, debug, initialProjectNodes, initialProjectProfileData, profile, env, }) {
        this._targetProjectAccountId = targetProjectAccountId;
        this._targetTestingAccountId = targetTestingAccountId;
        this._profile = profile;
        this._projectConfig = projectConfig;
        this._projectDir = projectDir;
        this._projectData = projectData;
        this._debug = debug || false;
        this._projectNodes = initialProjectNodes;
        this._projectNodesAtLastDeploy = initialProjectNodes;
        this._projectProfileData = initialProjectProfileData;
        this._env = env;
        this._appData = {};
        this._devServerMessage = LOCAL_DEV_SERVER_MESSAGE_TYPES.INITIAL;
        this._uploadWarnings = new Set();
        this._devServersStarted = false;
        this._listeners = {};
    }
    runListeners(key) {
        if (this._listeners[key] && this._listeners[key].length) {
            this._listeners[key].forEach(listener => listener(this[key]));
        }
    }
    get targetProjectAccountId() {
        return this._targetProjectAccountId;
    }
    get targetTestingAccountId() {
        return this._targetTestingAccountId;
    }
    get profile() {
        return this._profile;
    }
    get projectConfig() {
        return structuredClone(this._projectConfig);
    }
    get projectDir() {
        return this._projectDir;
    }
    get debug() {
        return this._debug;
    }
    get projectNodes() {
        return structuredClone(this._projectNodes);
    }
    set projectNodes(nodes) {
        this._projectNodes = nodes;
        this.runListeners('projectNodes');
    }
    get projectProfileData() {
        return structuredClone(this._projectProfileData);
    }
    set projectProfileData(profileData) {
        this._projectProfileData = profileData;
    }
    get projectNodesAtLastDeploy() {
        return structuredClone(this._projectNodesAtLastDeploy);
    }
    set projectNodesAtLastDeploy(nodes) {
        this._projectNodesAtLastDeploy = nodes;
    }
    get projectData() {
        return structuredClone(this._projectData);
    }
    get projectId() {
        return this.projectData.id;
    }
    set projectData(projectData) {
        this._projectData = projectData;
    }
    get env() {
        return this._env;
    }
    get appData() {
        return structuredClone(this._appData);
    }
    getAppDataByUid(uid) {
        return { ...this._appData[uid] };
    }
    setAppDataForUid(uid, appData) {
        this._appData[uid] = appData;
        this.runListeners('appData');
    }
    get devServerMessage() {
        return this._devServerMessage;
    }
    set devServerMessage(message) {
        this._devServerMessage = message;
        this.runListeners('devServerMessage');
    }
    get uploadWarnings() {
        return this._uploadWarnings;
    }
    get devServersStarted() {
        return this._devServersStarted;
    }
    set devServersStarted(started) {
        this._devServersStarted = started;
        this.runListeners('devServersStarted');
    }
    addUploadWarning(warning) {
        this.uploadWarnings.add(warning);
        this.runListeners('uploadWarnings');
    }
    clearUploadWarnings() {
        this.uploadWarnings.clear();
        this.runListeners('uploadWarnings');
    }
    addListener(key, listener) {
        if (!this._listeners[key]) {
            this._listeners[key] = [];
        }
        this._listeners[key].push(listener);
        listener(this[key]);
    }
    removeListener(key, listener) {
        if (this._listeners[key]) {
            this._listeners[key].splice(this._listeners[key].indexOf(listener), 1);
        }
    }
}
export default LocalDevState;
