// @hubspot/cli/lib/projects/localDev/LocalDevLogger.js
import { hasLocalStateFlag } from '@hubspot/local-dev-lib/config';
import { getConfigDefaultAccountIfExists } from '@hubspot/local-dev-lib/config';
import { uiLogger } from '../../ui/logger.js';
import { uiLine, uiAccountDescription, uiCommandReference, } from '../../ui/index.js';
import { lib } from '../../../lang/en.js';
import SpinniesManager from '../../ui/SpinniesManager.js';
import { logError } from '../../errorHandlers/index.js';
import { isAutoOpenBrowserEnabled } from '../../configOptions.js';
import { CONFIG_LOCAL_STATE_FLAGS } from '../../constants.js';
class LocalDevLogger {
    state;
    mostRecentUploadWarning;
    constructor(state) {
        this.state = state;
        this.mostRecentUploadWarning = null;
    }
    logUploadInstructions(warning) {
        uiLogger.log('');
        uiLogger.warn(warning);
        uiLogger.log('');
        uiLogger.log(lib.LocalDevManager.uploadWarning.instructionsHeader);
        uiLogger.log(lib.LocalDevManager.uploadWarning.stopDev);
        uiLogger.log(lib.LocalDevManager.uploadWarning.runUpload(this.getUploadCommand()));
        uiLogger.log(lib.LocalDevManager.uploadWarning.restartDev);
    }
    handleError(e, langFunction) {
        if (this.state.debug) {
            uiLogger.error(e instanceof Error ? e.message : String(e));
        }
        uiLogger.error(langFunction(e instanceof Error ? e.message : ''));
    }
    getUploadCommand() {
        const currentDefaultAccount = getConfigDefaultAccountIfExists();
        return this.state.targetProjectAccountId !==
            currentDefaultAccount?.accountId
            ? uiCommandReference(`hs project upload --account=${this.state.targetProjectAccountId}`)
            : uiCommandReference('hs project upload');
    }
    uploadWarning() {
        // At the moment, there is only one additional warning. We may need to do this in a
        // more robust way in the future
        const additionalWarnings = Array.from(this.state.uploadWarnings).join('\n\n');
        const warning = `${lib.LocalDevManager.uploadWarning.defaultWarning} ${additionalWarnings}`;
        // Avoid logging the warning to the console if it is currently the most
        // recently logged warning. We do not want to spam the console with the same message.
        if (warning !== this.mostRecentUploadWarning) {
            this.logUploadInstructions(warning);
            this.mostRecentUploadWarning = warning;
        }
    }
    missingComponentsWarning(components) {
        const warning = lib.LocalDevManager.uploadWarning.missingComponents(components.join(', '));
        if (warning !== this.mostRecentUploadWarning) {
            this.logUploadInstructions(warning);
            this.mostRecentUploadWarning = warning;
        }
    }
    fileChangeError(e) {
        this.handleError(e, lib.LocalDevManager.devServer.fileChangeError);
    }
    devServerSetupError(e) {
        this.handleError(e, lib.LocalDevManager.devServer.setupError);
    }
    devServerStartError(e) {
        this.handleError(e, lib.LocalDevManager.devServer.startError);
    }
    devServerCleanupError(e) {
        this.handleError(e, lib.LocalDevManager.devServer.cleanupError);
    }
    resetSpinnies() {
        SpinniesManager.stopAll();
        SpinniesManager.init();
    }
    startupMessage() {
        if (!this.state.debug) {
            console.clear();
        }
        uiLogger.log(lib.LocalDevManager.headerMessage);
        uiLogger.log(lib.LocalDevManager.learnMoreLocalDevServer);
        uiLogger.log('');
        uiLogger.log(lib.LocalDevManager.running(this.state.projectConfig.name, uiAccountDescription(this.state.targetProjectAccountId)));
        uiLogger.log(lib.LocalDevManager.viewProjectLink(this.state.projectConfig.name, this.state.targetProjectAccountId));
        const showWelcomeScreen = !hasLocalStateFlag(CONFIG_LOCAL_STATE_FLAGS.LOCAL_DEV_UI_WELCOME);
        if (!isAutoOpenBrowserEnabled()) {
            uiLogger.log(lib.LocalDevManager.viewLocalDevUILink(this.state.targetTestingAccountId, showWelcomeScreen));
        }
        else {
            uiLogger.log('');
            uiLogger.log(lib.LocalDevManager.localDevUIAutoMessage(this.state.targetTestingAccountId, showWelcomeScreen));
        }
        uiLogger.log('');
        uiLogger.log(lib.LocalDevManager.quitHelper);
        uiLine();
        uiLogger.log('');
    }
    cleanupStart() {
        SpinniesManager.add('cleanupMessage', {
            text: lib.LocalDevManager.exitingStart,
        });
    }
    cleanupError() {
        SpinniesManager.fail('cleanupMessage', {
            text: lib.LocalDevManager.exitingFail,
        });
    }
    cleanupSuccess() {
        SpinniesManager.succeed('cleanupMessage', {
            text: lib.LocalDevManager.exitingSucceed,
        });
    }
    uploadInitiated() {
        uiLogger.log(lib.LocalDevProcess.uploadInitiated);
    }
    deployInitiated() {
        uiLogger.log(lib.LocalDevProcess.deployInitiated);
    }
    projectConfigMismatch() {
        uiLogger.log(lib.LocalDevProcess.projectConfigMismatch);
    }
    uploadError(error) {
        uiLogger.log('');
        logError(error);
        uiLogger.log(lib.LocalDevProcess.uploadFailed);
        uiLogger.log('');
    }
    uploadSuccess() {
        uiLogger.log('');
        uiLogger.log(lib.LocalDevProcess.uploadSuccess);
        uiLine();
        uiLogger.log('');
    }
    uploadSuccessAutoDeployDisabled() {
        uiLogger.warn(lib.LocalDevProcess.uploadSuccessAutoDeployDisabled);
        uiLine();
        uiLogger.log('');
    }
    deployError(error) {
        uiLogger.log('');
        if (error) {
            logError(error);
        }
        uiLogger.log(lib.LocalDevProcess.deployFailed);
        uiLogger.log('');
    }
    deploySuccess() {
        uiLogger.log('');
        uiLogger.log(lib.LocalDevProcess.deploySuccess);
        uiLine();
        uiLogger.log('');
    }
    monitorConsoleOutput() {
        const originalStdoutWrite = process.stdout.write.bind(process.stdout);
        function customStdoutWrite(chunk, encoding, callback) {
            // Reset the most recently logged warning
            if (this.mostRecentUploadWarning) {
                this.mostRecentUploadWarning = null;
            }
            if (typeof encoding === 'function') {
                return originalStdoutWrite(chunk, callback);
            }
            return originalStdoutWrite(chunk, encoding, callback);
        }
        customStdoutWrite.bind(this);
        process.stdout.write = customStdoutWrite;
    }
}
export default LocalDevLogger;
