// @hubspot/cli/lib/projects/localDev/UIExtensionsDevModeInterface.js
import { requestPorts } from '@hubspot/local-dev-lib/portManager';
import { logger } from '@hubspot/local-dev-lib/logger';
import { DevModeUnifiedInterface as UIEDevModeInterface } from '@hubspot/ui-extensions-dev-server';
import { getHubSpotApiOrigin, getHubSpotWebsiteOrigin, } from '@hubspot/local-dev-lib/urls';
class UIExtensionsDevModeInterface {
    localDevState;
    constructor(options) {
        this.localDevState = options.localDevState;
    }
    async setup() {
        return UIEDevModeInterface.setup({
            // @ts-expect-error TODO: reconcile types between CLI and UIE Dev Server
            components: this.localDevState.projectNodes,
            profileData: this.localDevState.projectProfileData,
            logger,
            urls: {
                api: getHubSpotApiOrigin(this.localDevState.env),
                web: getHubSpotWebsiteOrigin(this.localDevState.env),
            },
        });
    }
    async start() {
        return UIEDevModeInterface.start({
            accountId: this.localDevState.targetTestingAccountId,
            // @ts-expect-error TODO: reconcile types between CLI and UIE Dev Server
            projectConfig: this.localDevState.projectConfig,
            requestPorts,
        });
    }
    async fileChange(filePath, event) {
        return UIEDevModeInterface.fileChange(filePath, event);
    }
    async cleanup() {
        return UIEDevModeInterface.cleanup();
    }
}
export default UIExtensionsDevModeInterface;
