// @hubspot/project-parsing-lib/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProjectThemeDetails = exports.migrateThemes = exports.getUnsupportedTypeError = exports.getFailedToFetchSchemasError = exports.getMissingRequiredFieldError = exports.getMissingAccountIdError = exports.getMissingConfigError = exports.getMissingTypeError = exports.getInvalidJsonError = exports.AjvErrorKeyword = exports.hsProjectJsonFilename = exports.metafileExtension = exports.createAjvInstance = exports.getIntermediateRepresentationSchema = exports.mapToInternalType = exports.mapToUserFriendlyName = exports.coerceToValidUid = exports.validateUid = exports.migrate = exports.projectContainsHsMetaFiles = exports.getAllHsProfiles = exports.loadHsProfileFile = exports.getHsProfileFilename = exports.isTranslationError = void 0;
exports.translate = translate;
exports.translateForLocalDev = translateForLocalDev;
const files_1 = require("./lib/files");
const validation_1 = require("./lib/validation");
const transform_1 = require("./lib/transform");
const copy_1 = require("./lang/copy");
const localDev_1 = require("./lib/localDev");
const defaultOptions = {
    skipValidation: false,
};
async function translate(translationContext, translationOptions = defaultOptions) {
    const { skipValidation } = translationOptions;
    const metafileContents = await (0, files_1.loadHsMetaFiles)(translationContext);
    if (metafileContents.length === 0) {
        throw new Error(copy_1.errorMessages.project.noHsMetaFiles);
    }
    let hsProfileContents;
    if (translationOptions.profile) {
        hsProfileContents = (0, files_1.loadHsProfileFile)(translationContext.projectSourceDir, translationOptions.profile);
    }
    const transformations = (0, transform_1.transform)(metafileContents, translationContext, hsProfileContents);
    const intermediateRepresentation = (0, transform_1.getIntermediateRepresentation)(transformations, hsProfileContents, skipValidation);
    // Remove once extensions and serverless functions are supported
    if (!skipValidation) {
        await (0, validation_1.validateIntermediateRepresentation)(intermediateRepresentation, transformations, translationContext);
    }
    return intermediateRepresentation;
}
async function translateForLocalDev(translationContext, translationOptions) {
    const metafileContents = await (0, files_1.loadHsMetaFiles)(translationContext);
    if (metafileContents.length === 0) {
        throw new Error(copy_1.errorMessages.project.noHsMetaFiles);
    }
    let hsProfileContents;
    if (translationOptions?.profile) {
        hsProfileContents = (0, files_1.loadHsProfileFile)(translationContext.projectSourceDir, translationOptions.profile);
    }
    const transformation = (0, transform_1.transform)(metafileContents, translationContext, hsProfileContents);
    const baseIntermediateRepresentation = (0, transform_1.getIntermediateRepresentation)(transformation, hsProfileContents, true);
    const parsingErrors = (0, transform_1.getParsingErrors)(transformation);
    const projectNodes = (0, localDev_1.getLocalDevProjectNodes)(baseIntermediateRepresentation.intermediateNodesIndexedByUid, translationContext.projectSourceDir, translationOptions?.projectNodesAtLastUpload);
    let projectNodesWithErrors = {};
    if (translationOptions?.projectNodesAtLastUpload) {
        projectNodesWithErrors = (0, localDev_1.getRemovedNodesAndNodesWithErrors)(baseIntermediateRepresentation.intermediateNodesIndexedByUid, translationOptions.projectNodesAtLastUpload, parsingErrors);
    }
    const profileData = (0, localDev_1.getLocalDevProfileData)(baseIntermediateRepresentation.profileData?.vars.profileVariables);
    return {
        intermediateNodesIndexedByUid: {
            ...projectNodes,
            ...projectNodesWithErrors,
        },
        profileData,
    };
}
var errors_1 = require("./lib/errors");
Object.defineProperty(exports, "isTranslationError", { enumerable: true, get: function () { return errors_1.isTranslationError; } });
var profiles_1 = require("./lib/profiles");
Object.defineProperty(exports, "getHsProfileFilename", { enumerable: true, get: function () { return profiles_1.getHsProfileFilename; } });
var files_2 = require("./lib/files");
Object.defineProperty(exports, "loadHsProfileFile", { enumerable: true, get: function () { return files_2.loadHsProfileFile; } });
Object.defineProperty(exports, "getAllHsProfiles", { enumerable: true, get: function () { return files_2.getAllHsProfiles; } });
Object.defineProperty(exports, "projectContainsHsMetaFiles", { enumerable: true, get: function () { return files_2.projectContainsHsMetaFiles; } });
var migrate_1 = require("./lib/migrate");
Object.defineProperty(exports, "migrate", { enumerable: true, get: function () { return migrate_1.migrate; } });
var uid_1 = require("./lib/uid");
Object.defineProperty(exports, "validateUid", { enumerable: true, get: function () { return uid_1.validateUid; } });
Object.defineProperty(exports, "coerceToValidUid", { enumerable: true, get: function () { return uid_1.coerceToValidUid; } });
var transform_2 = require("./lib/transform");
Object.defineProperty(exports, "mapToUserFriendlyName", { enumerable: true, get: function () { return transform_2.mapToUserFriendlyName; } });
Object.defineProperty(exports, "mapToInternalType", { enumerable: true, get: function () { return transform_2.mapToInternalType; } });
var schemas_1 = require("./lib/schemas");
Object.defineProperty(exports, "getIntermediateRepresentationSchema", { enumerable: true, get: function () { return schemas_1.getIntermediateRepresentationSchema; } });
var validation_2 = require("./lib/validation");
Object.defineProperty(exports, "createAjvInstance", { enumerable: true, get: function () { return validation_2.createAjvInstance; } });
var constants_1 = require("./lib/constants");
Object.defineProperty(exports, "metafileExtension", { enumerable: true, get: function () { return constants_1.metafileExtension; } });
Object.defineProperty(exports, "hsProjectJsonFilename", { enumerable: true, get: function () { return constants_1.hsProjectJsonFilename; } });
var errors_2 = require("./lib/errors");
Object.defineProperty(exports, "AjvErrorKeyword", { enumerable: true, get: function () { return errors_2.AjvErrorKeyword; } });
var copy_2 = require("./lang/copy");
Object.defineProperty(exports, "getInvalidJsonError", { enumerable: true, get: function () { return copy_2.getInvalidJsonError; } });
Object.defineProperty(exports, "getMissingTypeError", { enumerable: true, get: function () { return copy_2.getMissingTypeError; } });
Object.defineProperty(exports, "getMissingConfigError", { enumerable: true, get: function () { return copy_2.getMissingConfigError; } });
Object.defineProperty(exports, "getMissingAccountIdError", { enumerable: true, get: function () { return copy_2.getMissingAccountIdError; } });
Object.defineProperty(exports, "getMissingRequiredFieldError", { enumerable: true, get: function () { return copy_2.getMissingRequiredFieldError; } });
Object.defineProperty(exports, "getFailedToFetchSchemasError", { enumerable: true, get: function () { return copy_2.getFailedToFetchSchemasError; } });
Object.defineProperty(exports, "getUnsupportedTypeError", { enumerable: true, get: function () { return copy_2.getUnsupportedTypeError; } });
var migrateThemes_1 = require("./lib/migrateThemes");
Object.defineProperty(exports, "migrateThemes", { enumerable: true, get: function () { return migrateThemes_1.migrateThemes; } });
Object.defineProperty(exports, "getProjectThemeDetails", { enumerable: true, get: function () { return migrateThemes_1.getProjectThemeDetails; } });
