// @hubspot/cli/commands/project/list.js
import { trackCommandUsage } from '../../lib/usageTracking.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
import { fetchProjects } from '@hubspot/local-dev-lib/api/projects';
import { logError } from '../../lib/errorHandlers/index.js';
import { EXIT_CODES } from '../../lib/enums/exitCodes.js';
import { getTableContents, getTableHeader } from '../../lib/ui/table.js';
import { uiLogger } from '../../lib/ui/logger.js';
import { commands } from '../../lang/en.js';
const command = ['list', 'ls'];
const describe = commands.project.list.describe;
async function getProjectData(accountId) {
    try {
        const { data: projects } = await fetchProjects(accountId);
        return projects.results;
    }
    catch (e) {
        logError(e);
        process.exit(EXIT_CODES.ERROR);
    }
}
function formatProjectsAsTableRows(projects) {
    const projectListData = [];
    projects.forEach(project => {
        projectListData.push([
            project.name,
            project.latestBuild ? project.latestBuild.platformVersion : '',
        ]);
    });
    return projectListData;
}
async function handler(args) {
    const { derivedAccountId } = args;
    trackCommandUsage('projects-list', undefined, derivedAccountId);
    const projectData = await getProjectData(derivedAccountId);
    if (projectData.length === 0) {
        uiLogger.error(commands.project.list.errors.noProjectsFound(derivedAccountId));
        process.exit(EXIT_CODES.ERROR);
    }
    const projectListData = formatProjectsAsTableRows(projectData);
    projectListData.unshift(getTableHeader([
        commands.project.list.labels.name,
        commands.project.list.labels.platformVersion,
    ]));
    uiLogger.log(commands.project.list.projects);
    uiLogger.log(getTableContents(projectListData, { border: { bodyLeft: '  ' } }));
}
function projectListBuilder(yargs) {
    yargs.example([['$0 project list']]);
    return yargs;
}
const builder = makeYargsBuilder(projectListBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
});
const projectListCommand = {
    command,
    describe,
    handler,
    builder,
};
export default projectListCommand;
