// @hubspot/cli/commands/project/updateDeps.js
import { updatePackages, getProjectPackageJsonLocations, } from '../../lib/dependencyManagement.js';
import { EXIT_CODES } from '../../lib/enums/exitCodes.js';
import { getProjectConfig } from '../../lib/projects/config.js';
import { promptUser } from '../../lib/prompts/promptUtils.js';
import path from 'path';
import { commands } from '../../lang/en.js';
import { uiLogger } from '../../lib/ui/logger.js';
import { trackCommandUsage } from '../../lib/usageTracking.js';
import { logError } from '../../lib/errorHandlers/index.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
const command = 'update-deps [packages..]';
const describe = commands.project.updateDeps.help.describe;
async function handler(args) {
    const { derivedAccountId, packages } = args;
    try {
        trackCommandUsage('project-update-deps', undefined, derivedAccountId);
        const projectConfig = await getProjectConfig();
        if (!projectConfig || !projectConfig.projectDir) {
            uiLogger.error(commands.project.updateDeps.noProjectConfig);
            return process.exit(EXIT_CODES.ERROR);
        }
        const { projectDir } = projectConfig;
        let installLocations = await getProjectPackageJsonLocations();
        if (packages) {
            const { selectedInstallLocations } = await promptUser([
                {
                    name: 'selectedInstallLocations',
                    type: 'checkbox',
                    when: () => packages && packages.length > 0,
                    message: commands.project.updateDeps.installLocationPrompt,
                    choices: installLocations.map(dir => ({
                        name: path.relative(projectDir, dir),
                        value: dir,
                    })),
                    validate: (choices) => {
                        if (choices === undefined || choices.length === 0) {
                            return commands.project.updateDeps.installLocationPromptRequired;
                        }
                        return true;
                    },
                },
            ]);
            if (selectedInstallLocations) {
                installLocations = selectedInstallLocations;
            }
        }
        await updatePackages({
            packages,
            installLocations,
        });
    }
    catch (e) {
        logError(e);
        return process.exit(EXIT_CODES.ERROR);
    }
}
function projectUpdateDepsBuilder(yargs) {
    yargs.example([
        [
            '$0 project update-deps',
            commands.project.updateDeps.help.updateAppDepsExample,
        ],
        [
            '$0 project update-deps dependency1 dependency2',
            commands.project.updateDeps.help.updateDepToSubComponentExample,
        ],
    ]);
    return yargs;
}
const builder = makeYargsBuilder(projectUpdateDepsBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
});
const projectUpdateDepsCommand = {
    command,
    describe,
    handler,
    builder,
};
export default projectUpdateDepsCommand;
