// @hubspot/cli/commands/customObject/createSchema.js
import { uiLogger } from '../../lib/ui/logger.js';
import { getConfigAccountEnvironment } from '@hubspot/local-dev-lib/config';
import { getAbsoluteFilePath } from '@hubspot/local-dev-lib/path';
import { createObjectSchema } from '@hubspot/local-dev-lib/api/customObjects';
import { getHubSpotWebsiteOrigin } from '@hubspot/local-dev-lib/urls';
import { logError } from '../../lib/errorHandlers/index.js';
import { checkAndConvertToJson } from '../../lib/validation.js';
import { trackCommandUsage } from '../../lib/usageTracking.js';
import { commands } from '../../lang/en.js';
import { EXIT_CODES } from '../../lib/enums/exitCodes.js';
import { isSchemaDefinition } from '../../lib/customObject.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
const command = 'create-schema';
const describe = commands.customObject.subcommands.createSchema.describe;
async function handler(args) {
    const { path, derivedAccountId } = args;
    trackCommandUsage('custom-object-schema-create', {}, derivedAccountId);
    const filePath = getAbsoluteFilePath(path);
    const schemaJson = checkAndConvertToJson(filePath);
    if (!isSchemaDefinition(schemaJson)) {
        uiLogger.error(commands.customObject.subcommands.createSchema.errors.invalidSchema);
        process.exit(EXIT_CODES.ERROR);
    }
    try {
        const { data } = await createObjectSchema(derivedAccountId, schemaJson);
        uiLogger.success(commands.customObject.subcommands.createSchema.success.schemaViewable(`${getHubSpotWebsiteOrigin(getConfigAccountEnvironment(derivedAccountId))}/contacts/${derivedAccountId}/objects/${data.objectTypeId}`));
    }
    catch (e) {
        logError(e, { accountId: derivedAccountId });
        uiLogger.error(commands.customObject.subcommands.createSchema.errors.creationFailed(path));
    }
}
function schemaCreateBuilder(yargs) {
    yargs.option('path', {
        describe: commands.customObject.subcommands.createSchema.options.definition
            .describe,
        type: 'string',
        required: true,
    });
    return yargs;
}
const builder = makeYargsBuilder(schemaCreateBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
    useTestingOptions: true,
});
const createSchemaCommand = {
    command,
    describe,
    handler,
    builder,
};
export default createSchemaCommand;
