// @hubspot/cli/commands/customObject/fetchAllSchemas.js
import { uiLogger } from '../../lib/ui/logger.js';
import { downloadSchemas, getResolvedPath, } from '@hubspot/local-dev-lib/customObjects';
import { inputPrompt } from '../../lib/prompts/promptUtils.js';
import { trackCommandUsage } from '../../lib/usageTracking.js';
import { commands } from '../../lang/en.js';
import { logSchemas } from '../../lib/schema.js';
import { logError } from '../../lib/errorHandlers/index.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
const command = 'fetch-all-schemas [dest]';
const describe = commands.customObject.subcommands.fetchAllSchemas.describe;
async function handler(args) {
    const { derivedAccountId, dest: providedDest } = args;
    trackCommandUsage('custom-object-schema-fetch-all', {}, derivedAccountId);
    try {
        const dest = providedDest ||
            (await inputPrompt(commands.customObject.subcommands.fetchAllSchemas.inputDest));
        const schemas = await downloadSchemas(derivedAccountId, dest);
        logSchemas(schemas);
        uiLogger.success(commands.customObject.subcommands.fetchAllSchemas.success.fetch(getResolvedPath(dest)));
    }
    catch (e) {
        logError(e);
        uiLogger.error(commands.customObject.subcommands.fetchAllSchemas.errors.fetch);
    }
}
function schemaFetchAllBuilder(yargs) {
    yargs
        .example([
        [
            '$0 custom-object fetch-all-schemas',
            commands.customObject.subcommands.fetchAllSchemas.examples.default,
        ],
        [
            '$0 custom-object fetch-all-schemas my/folder',
            commands.customObject.subcommands.fetchAllSchemas.examples.specifyPath,
        ],
    ])
        .positional('dest', {
        describe: commands.customObject.subcommands.fetchAllSchemas.positionals.dest
            .describe,
        type: 'string',
    });
    return yargs;
}
const builder = makeYargsBuilder(schemaFetchAllBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const fetchAllSchemasCommand = {
    command,
    describe,
    handler,
    builder,
};
export default fetchAllSchemasCommand;
