// @hubspot/cli/commands/customObject/listSchemas.js
import { uiLogger } from '../../lib/ui/logger.js';
import { logError } from '../../lib/errorHandlers/index.js';
import { trackCommandUsage } from '../../lib/usageTracking.js';
import { listSchemas } from '../../lib/schema.js';
import { commands } from '../../lang/en.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
const command = 'list-schemas';
const describe = commands.customObject.subcommands.listSchemas.describe;
async function handler(args) {
    const { derivedAccountId } = args;
    trackCommandUsage('custom-object-schema-list', {}, derivedAccountId);
    try {
        await listSchemas(derivedAccountId);
    }
    catch (e) {
        logError(e);
        uiLogger.error(commands.customObject.subcommands.listSchemas.errors.list);
    }
}
function schemaListBuilder(yargs) {
    return yargs;
}
const builder = makeYargsBuilder(schemaListBuilder, command, describe, {
    useGlobalOptions: true,
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const listSchemasCommand = {
    command,
    describe,
    handler,
    builder,
};
export default listSchemasCommand;
