// @hubspot/ui-extensions-dev-server/dist/lib/config.js
import fs from 'fs';
import path from 'path';
import { buildSourceId, getUrlSafeFileName } from "./utils.js";
export function loadConfigByPath(configPath) {
    const source = fs.readFileSync(configPath).toString();
    return JSON.parse(source);
}
export function validateCardConfig(config) {
    if (!config || typeof config !== 'object') {
        return new Error('Card config must be an object');
    }
    if (!('type' in config) || config.type !== 'crm-card') {
        return new Error('`type` must be "crm-card"');
    }
    if (!('data' in config) ||
        typeof config.data !== 'object' ||
        config.data === null) {
        return new Error('`data` must be an object');
    }
    const { data } = config;
    if (!('title' in data) || typeof data.title !== 'string') {
        return new Error('`data` must have a `title` string');
    }
    if (!('location' in data) || typeof data.location !== 'string') {
        return new Error('`data` must have a `location` string');
    }
    if (!('module' in data) ||
        typeof data.module !== 'object' ||
        data.module === null) {
        return new Error('`data` must have a `module` object');
    }
    if (!('file' in data.module) || typeof data.module.file !== 'string') {
        return new Error('`data.module` must have a `file` string');
    }
    if (!('objectTypes' in data) || !Array.isArray(data.objectTypes)) {
        return new Error('`data.module` must have an `objectTypes` array');
    }
    if (data.objectTypes.length === 0 ||
        data.objectTypes.some(({ name }) => typeof name !== 'string')) {
        return new Error('all `data.module.objectTypes` objects must have `name` strings');
    }
    return true;
}
export function loadExtensionConfig(appConfig, appPath) {
    const crmCardsSubConfigFiles = appConfig?.extensions?.crm?.cards;
    if (!crmCardsSubConfigFiles) {
        throw new Error("Unable to find extensions files, make sure the 'extensions.crm.cards' array is defined in the application configuration file");
    }
    const outputConfig = {};
    crmCardsSubConfigFiles.forEach((card) => {
        const cardConfigPath = path.join(appPath, card.file);
        try {
            const cardConfig = loadConfigByPath(cardConfigPath);
            const validation = validateCardConfig(cardConfig);
            if (validation !== true) {
                console.error(`[DevServer] Extension config ${cardConfigPath} is invalid: ${validation.message}`);
            }
            if (cardConfig && cardConfig.data) {
                const cardConfigDir = path.parse(cardConfigPath).dir;
                const entryPointPath = path.join(cardConfigDir, cardConfig.data?.module?.file);
                cardConfig.data.module.file = entryPointPath;
                const sourceId = buildSourceId(appConfig, cardConfig);
                outputConfig[sourceId || `${entryPointPath}-${cardConfig.data.location}`] = {
                    ...cardConfig,
                    output: getUrlSafeFileName(entryPointPath),
                    path: appPath,
                    extensionPath: path.parse(entryPointPath).dir,
                    extensionConfigPath: cardConfigPath,
                    data: {
                        ...cardConfig.data,
                        appName: appConfig.name,
                        sourceId,
                    },
                    appConfig,
                };
            }
        }
        catch (e) {
            throw new Error(`Unable to load ${cardConfigPath}`);
        }
    });
    return outputConfig;
}
export function validateProxyConfigKey(urlKey, logger, localConfigPath) {
    try {
        const url = new URL(urlKey);
        if (url.pathname !== '/') {
            logger.warn(`The key "${urlKey}" in "${localConfigPath}" is invalid, paths are not supported for keys`);
        }
    }
    catch (e) {
        logger.warn(`The key "${urlKey}" in "${localConfigPath}" is an invalid url`);
    }
}
export function validateProxyConfigValue(value, key, logger, localConfigPath) {
    try {
        // eslint-disable-next-line no-new
        new URL(value);
    }
    catch (e) {
        logger.warn(`The value "${value}" for key "${key}" in "${localConfigPath}" is an invalid url`);
    }
}
export function loadLocalConfig(appPath, logger) {
    const localConfigFilename = 'local.json';
    const localConfigPath = path.join(appPath, localConfigFilename);
    if (!fs.existsSync(localConfigPath)) {
        return undefined;
    }
    try {
        const localConfig = JSON.parse(fs.readFileSync(localConfigPath).toString());
        const { proxy = {} } = localConfig;
        Object.entries(proxy).forEach((entry) => {
            const [key, value] = entry;
            validateProxyConfigKey(key, logger, localConfigFilename);
            validateProxyConfigValue(value, key, logger, localConfigFilename);
        });
        return localConfig;
    }
    catch (e) {
        logger.error(`Error loading and parsing ${localConfigPath}, ${e}`);
        return undefined;
    }
}
