// @hubspot/ui-extensions-dev-server/dist/lib/constants.js
export const OUTPUT_DIR = 'dist';
export const MANIFEST_FILE = 'manifest.json';
export const EXPRESS_SERVER_ID = 'ui-extensions-dev-server';
export const EXPRESS_DEFAULT_PORT = 5173;
export const ROLLUP_OPTIONS = {
    // Deps to exclude from the bundle
    external: ['react', 'react-dom', '@remote-ui/react'],
    output: {
        // Maps libs to the variables to be injected via the window
        globals: {
            react: 'React',
            '@remote-ui/react': 'RemoteUI',
        },
        extend: true,
    },
};
export const EXTENSIONS_MESSAGE_VERSION = 2;
export const WEBSOCKET_MESSAGE_VERSION = 1;
export const PROXY_CAPABILITY = 'app-backend-proxy-server';
export const SERVER_CAPABILITIES = [
    // Supports running app functions locally
    'app-functions-local-dev',
    'source-id',
    'account-id',
];
export const PLATFORM_VERSION = {
    V20231: '2023.1',
    V20232: '2023.2',
    V20251: '2025.1',
    V20252: '2025.2',
    UNSTABLE: 'unstable',
};
export const PUBLIC_APP = 'public-app';
export const PRIVATE_APP = 'private-app';
export const CARD_EXTENSION = 'CARD';
export const SETTINGS_EXTENSION = 'SETTINGS';
export const PAGE_EXTENSION = 'PAGE';
export const SUPPORTED_APP_TYPES = [PUBLIC_APP, PRIVATE_APP];
export const SUPPORTED_EXTENSION_TYPES = [
    CARD_EXTENSION,
    SETTINGS_EXTENSION,
    PAGE_EXTENSION,
];
