// @hubspot/ui-extensions-dev-server/dist/lib/DevServerState.js
import { OUTPUT_DIR } from "./constants.js";
import path from 'path';
export class DevServerState {
    _expressPort;
    _functionsConfig;
    _localDevUrlMapping;
    _outputDir;
    _appPath;
    _extensionsMetadata;
    _portalId;
    _mutableState = {};
    logger;
    appConfig;
    constructor({ localDevUrlMapping, extensionConfigs, accountId, expressPort, platformVersion, logger, urls, appConfig, }) {
        if (!extensionConfigs) {
            throw new Error('Unable to load the required extension configuration files');
        }
        const extensionsMetadata = [];
        extensionConfigs.forEach((config) => {
            const { appName, title, sourceId } = config.data;
            extensionsMetadata.push({
                config,
                baseMessage: {
                    appName,
                    title,
                    sourceId,
                    callback: `http://hslocal.net:${expressPort}/${config.output}`,
                    portalId: accountId,
                },
            });
        });
        this._expressPort = expressPort;
        this._extensionsMetadata = extensionsMetadata;
        this._appPath = extensionConfigs[0].path;
        this._portalId = accountId;
        this._outputDir = path.join(this._appPath, OUTPUT_DIR);
        // Pass options from the CLI for running app functions locally
        this._functionsConfig = {
            app: { path: this._appPath },
            accountId,
            platformVersion,
            hubspotApiOrigin: urls.api,
            hubspotWebsiteOrigin: urls.web,
        };
        this._localDevUrlMapping = localDevUrlMapping;
        this.logger = logger;
        this.appConfig = appConfig;
        Object.freeze(this);
    }
    get portalId() {
        return this._portalId;
    }
    get expressPort() {
        return this._expressPort;
    }
    get extensionsMetadata() {
        return this._extensionsMetadata;
    }
    get functionsConfig() {
        return this._functionsConfig;
    }
    get outputDir() {
        return this._outputDir;
    }
    get appPath() {
        return this._appPath;
    }
    get localDevUrlMapping() {
        return this._localDevUrlMapping;
    }
    get extensionsWebSocket() {
        return this._mutableState.extensionsWebSocket;
    }
    set extensionsWebSocket(ws) {
        this._mutableState.extensionsWebSocket = ws;
    }
    getExtensionsWebSocket() {
        if (!this._mutableState.extensionsWebSocket) {
            throw new Error('ExtensionsWebSocket not initialized. Server must be started first.');
        }
        return this._mutableState.extensionsWebSocket;
    }
    isPublicApp() {
        return !!this.appConfig?.isPublicApp;
    }
    setWebSocketSetupCallback(callback) {
        this._mutableState.webSocketSetupCallback = callback;
    }
    triggerWebSocketSetup() {
        if (this._mutableState.webSocketSetupCallback) {
            this._mutableState.webSocketSetupCallback();
            this._mutableState.webSocketSetupCallback = undefined;
        }
    }
}
