// @hubspot/ui-extensions-dev-server/dist/lib/extensionsService.js
import path from 'path';
import { EXTENSIONS_MESSAGE_VERSION } from "./constants.js";
import { loadManifest } from "./utils.js";
class ExtensionsService {
    endpoint;
    constructor() {
        this.endpoint = '/extensions';
    }
    add(server, devServerState, capabilities) {
        server.get(this.endpoint, this.generateExtensionsHandler(devServerState, capabilities));
        return [this.endpoint];
    }
    generateExtensionsHandler(devServerState, capabilities = []) {
        return function extensionsHandler(_req, res) {
            try {
                const extensions = devServerState.extensionsMetadata.map((metadata) => {
                    const { baseMessage } = metadata;
                    const output = path.parse(baseMessage.callback).name;
                    return {
                        ...baseMessage,
                        manifest: loadManifest(devServerState.outputDir, output),
                    };
                });
                const response = {
                    websocket: `ws://localhost:${devServerState.expressPort}`,
                    version: EXTENSIONS_MESSAGE_VERSION,
                    capabilities,
                    portalId: devServerState.portalId,
                    extensions,
                    ...(devServerState.functionsConfig.platformVersion && {
                        platformVersion: devServerState.functionsConfig.platformVersion,
                    }),
                };
                res.status(200).json(response);
            }
            catch (e) {
                devServerState.logger.error(`Error in /extensions endpoint: ${e}`);
                res.status(500).json({
                    message: 'Unable to determine which extensions are running',
                });
            }
        };
    }
}
export default new ExtensionsService();
