// @hubspot/ui-extensions-dev-server/dist/lib/server.js
import express from 'express';
import cors from 'cors';
import { PROXY_CAPABILITY, SERVER_CAPABILITIES } from "./constants.js";
import extensionsService from "./extensionsService.js";
import { AppFunctionExecutionService, AppProxyService, } from '@hubspot/app-functions-dev-server';
import { extractAllowedUrls } from "./utils.js";
import { ExtensionsWebSocket } from "./ExtensionsWebSocket.js";
function listen(app, port) {
    return new Promise((resolve, reject) => {
        const server = app
            .listen({ port }, () => {
            resolve(server);
        })
            .on('error', (err) => {
            reject(err);
        });
    });
}
async function startDevServer({ devServerState, viteDevServer, }) {
    const app = express();
    // Setup middleware
    app.use(cors());
    app.use(express.static(devServerState.outputDir));
    const capabilities = [...SERVER_CAPABILITIES];
    const middlewares = [];
    if (!devServerState.isPublicApp()) {
        middlewares.push(AppFunctionExecutionService({
            ...devServerState.functionsConfig,
            logger: devServerState.logger,
        }));
        devServerState.logger.info(`Serving app functions locally (platform version ${devServerState.functionsConfig.platformVersion})`);
    }
    if (devServerState.localDevUrlMapping &&
        Object.keys(devServerState.localDevUrlMapping).length > 0) {
        devServerState.logger.info('Proxy config discovered, enabling local proxy mode');
        middlewares.push(AppProxyService({
            localDevUrlMapping: devServerState.localDevUrlMapping,
            logger: devServerState.logger,
            accountId: devServerState.functionsConfig.accountId,
            allowedUrls: extractAllowedUrls(devServerState.appConfig),
        }));
        capabilities.push(PROXY_CAPABILITY);
    }
    if (middlewares.length > 0) {
        app.use('/api/crm-extensibility/execution/internal/v3', ...middlewares);
    }
    const endpointsAdded = extensionsService.add(app, devServerState, capabilities);
    const { expressPort } = devServerState;
    endpointsAdded.forEach((endpoint) => {
        devServerState.logger.debug(`Listening at http://hslocal.net:${expressPort}${endpoint}`);
    });
    // Vite middlewares needs to go last because it's greedy and will block other middleware
    app.use(viteDevServer.middlewares);
    let server;
    try {
        server = await listen(app, devServerState.expressPort);
    }
    catch (e) {
        if (e.code === 'EADDRINUSE') {
            throw new Error(`Port ${devServerState.expressPort} is already in use.`);
        }
        throw new Error(e);
    }
    /**
     * We use a custom WebSocket server instead of Vite's built-in one because:
     * 1. We send HubSpot-specific extension metadata (not standard HMR messages)
     * 2. We need cross-origin connections from HubSpot iframes
     * 3. Our message format { event: 'start', ...metadata } doesn't match Vite's protocol
     */
    const wss = new ExtensionsWebSocket(server, devServerState);
    devServerState.extensionsWebSocket = wss;
    // Trigger the WebSocket setup that was deferred from the plugin's configureServer hook
    // This must happen after the WebSocket is initialized to avoid race conditions
    devServerState.triggerWebSocketSetup();
    devServerState.extensionsMetadata?.forEach((metadata) => {
        const { baseMessage } = metadata;
        devServerState.logger.debug(`Listening at ${baseMessage.callback}`);
    });
    return {
        httpServer: server,
        shutdown: async function shutdown() {
            try {
                await viteDevServer.pluginContainer.close();
                // Close WebSocket server first to reject new connections
                await devServerState.getExtensionsWebSocket().close();
                // Close HTTP server
                await new Promise((resolve) => {
                    server.close(() => resolve());
                });
                await viteDevServer.close();
            }
            catch (e) {
                devServerState.logger.debug(`Error shutting down ${e}`);
            }
            devServerState.logger.info('Extension dev server done cleaning up');
        },
    };
}
export default startDevServer;
