// @hubspot/cli/api/migrate.js
import { PLATFORM_VERSIONS, } from '@hubspot/local-dev-lib/constants/projects';
import { http } from '@hubspot/local-dev-lib/http';
const MIGRATIONS_API_PATH_V2 = 'dfs/migrations/v2';
export const CLI_UNMIGRATABLE_REASONS = {
    PART_OF_PROJECT_ALREADY: 'PART_OF_PROJECT_ALREADY',
};
export function isMigrationStatus(error) {
    return (typeof error === 'object' &&
        error !== null &&
        'id' in error &&
        'status' in error);
}
export async function listAppsForMigration(accountId, platformVersion) {
    return http.get(accountId, {
        url: `${MIGRATIONS_API_PATH_V2}/list-apps`,
        params: {
            platformVersion: mapPlatformVersionToEnum(platformVersion),
        },
    });
}
function mapPlatformVersionToEnum(platformVersion) {
    if (platformVersion === PLATFORM_VERSIONS.unstable) {
        return PLATFORM_VERSIONS.unstable.toUpperCase();
    }
    return `V${platformVersion.replace('.', '_')}`;
}
export async function initializeAppMigration(accountId, applicationId, platformVersion) {
    return http.post(accountId, {
        url: `${MIGRATIONS_API_PATH_V2}/migrations`,
        data: {
            applicationId,
            platformVersion: mapPlatformVersionToEnum(platformVersion),
        },
    });
}
export async function continueAppMigration(portalId, migrationId, componentUids, projectName) {
    return http.post(portalId, {
        url: `${MIGRATIONS_API_PATH_V2}/migrations/continue`,
        data: {
            migrationId,
            projectName,
            componentUids,
        },
    });
}
export async function checkMigrationStatusV2(accountId, id) {
    return http.get(accountId, {
        url: `${MIGRATIONS_API_PATH_V2}/migrations/${id}/status`,
    });
}
