// @hubspot/app-functions-dev-server/dist/services/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateTokenRefreshMessage = exports.generateTokenOpErrorMessage = exports.generateCreateTokenMessage = exports.generateTokenCachedMessage = exports.generateCahedTokeMessage = exports.generateTokensNotEnabledMessage = exports.generateMissingScopesMoreContextMessage = exports.generateTokensEnableMessage = exports.generateMissingScopesMessage = void 0;
function generateMissingScopesMessage(accountId) {
    return `Your account is missing the "App functions" scope, so some new features won’t work. To fix this, regenerate the Personal Access Key for account: ${accountId}`;
}
exports.generateMissingScopesMessage = generateMissingScopesMessage;
function generateTokensEnableMessage(accountId) {
    return `Private app user tokens are enabled for accountId: ${accountId}`;
}
exports.generateTokensEnableMessage = generateTokensEnableMessage;
function generateMissingScopesMoreContextMessage(accountId) {
    return `
    Heads up! Private app user tokens are now supported for serverless functions, but it looks like the account (ID: ${accountId}) is missing the required "App functions" scope.

    To fix this:
    1. Deactivate the existing Personal Access Key for this account.
    2. Generate a new Personal Access Key with the "App functions" scope enabled.
    3. Run "hs auth" and reauthenticate the account (ID: ${accountId}).

    Once updated, you won't need a "PRIVATE_APP_ACCESS_TOKEN" for local development.
  `;
}
exports.generateMissingScopesMoreContextMessage = generateMissingScopesMoreContextMessage;
function generateTokensNotEnabledMessage(accountId, appId) {
    return `Private app user tokens are not enabled for accountId ${accountId}, skipping call to fetch for appId ${appId}`;
}
exports.generateTokensNotEnabledMessage = generateTokensNotEnabledMessage;
function generateCahedTokeMessage(appId) {
    return `Using cached private app user token for appId ${appId}`;
}
exports.generateCahedTokeMessage = generateCahedTokeMessage;
function generateTokenCachedMessage(appId, expiresAt) {
    return `Token for appId ${appId} expiring at ${expiresAt} is in cache.`;
}
exports.generateTokenCachedMessage = generateTokenCachedMessage;
function generateCreateTokenMessage(appId) {
    return `Creating new private app user token for appId ${appId}`;
}
exports.generateCreateTokenMessage = generateCreateTokenMessage;
function generateTokenOpErrorMessage({ appId, accountId, messageDetail = '', operation = 'create', }) {
    return `Unable to ${operation} private app user token for appId ${appId} on accountId ${accountId}. ${messageDetail}`;
}
exports.generateTokenOpErrorMessage = generateTokenOpErrorMessage;
function generateTokenRefreshMessage(appId) {
    return `Refreshing private app user token for appId ${appId}`;
}
exports.generateTokenRefreshMessage = generateTokenRefreshMessage;
