// @hubspot/cli/commands/function/deploy.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../lib/ui/index.js';
import cmsFunctionDeployCommand from '../cms/function/deploy.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
import { commands } from '../../lang/en.js';
const command = 'deploy <path>';
const describe = uiDeprecatedTag(cmsFunctionDeployCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms function deploy');
    await cmsFunctionDeployCommand.handler(args);
}
function deprecatedFunctionDeployBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.function.subcommands.deploy.positionals.path
            .describe,
        type: 'string',
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(cmsFunctionDeployCommand.describe, 'hs cms function deploy');
const builder = makeYargsBuilder(deprecatedFunctionDeployBuilder, command, verboseDescribe, {
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedFunctionDeployCommand = {
    ...cmsFunctionDeployCommand,
    describe,
    handler,
    builder,
};
export default deprecatedFunctionDeployCommand;
