// @hubspot/cli/commands/function/list.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../lib/ui/index.js';
import cmsFunctionListCommand from '../cms/function/list.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
import { commands } from '../../lang/en.js';
const command = ['list', 'ls'];
const describe = uiDeprecatedTag(cmsFunctionListCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms function list');
    await cmsFunctionListCommand.handler(args);
}
function deprecatedFunctionListBuilder(yargs) {
    yargs.options({
        json: {
            describe: commands.cms.subcommands.function.subcommands.list.options.json
                .describe,
            type: 'boolean',
        },
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(cmsFunctionListCommand.describe, 'hs cms function list');
const builder = makeYargsBuilder(deprecatedFunctionListBuilder, command, verboseDescribe, {
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedFunctionListCommand = {
    ...cmsFunctionListCommand,
    describe,
    handler,
    builder,
};
export default deprecatedFunctionListCommand;
