// @hubspot/cli/commands/function/server.js
import { uiCommandRelocatedMessage, uiCommandRenamedDescription, uiDeprecatedTag, } from '../../lib/ui/index.js';
import cmsFunctionServerCommand from '../cms/function/server.js';
import { makeYargsBuilder } from '../../lib/yargsUtils.js';
import { commands } from '../../lang/en.js';
const command = 'server <path>';
const describe = uiDeprecatedTag(cmsFunctionServerCommand.describe, false);
async function handler(args) {
    uiCommandRelocatedMessage('hs cms function server');
    await cmsFunctionServerCommand.handler(args);
}
function deprecatedFunctionServerBuilder(yargs) {
    yargs.positional('path', {
        describe: commands.cms.subcommands.function.subcommands.server.positionals.path
            .describe,
        type: 'string',
    });
    yargs.options({
        port: {
            describe: commands.cms.subcommands.function.subcommands.server.options.port
                .describe,
            type: 'string',
            default: 5432,
        },
        contact: {
            describe: commands.cms.subcommands.function.subcommands.server.options.contact
                .describe,
            type: 'boolean',
            default: true,
        },
        watch: {
            describe: commands.cms.subcommands.function.subcommands.server.options.watch
                .describe,
            type: 'boolean',
            default: true,
        },
        'log-output': {
            describe: commands.cms.subcommands.function.subcommands.server.options.logOutput
                .describe,
            type: 'boolean',
            default: false,
        },
    });
    return yargs;
}
const verboseDescribe = uiCommandRenamedDescription(cmsFunctionServerCommand.describe, 'hs cms function server');
const builder = makeYargsBuilder(deprecatedFunctionServerBuilder, command, verboseDescribe, {
    useConfigOptions: true,
    useAccountOptions: true,
    useEnvironmentOptions: true,
});
const deprecatedFunctionServerCommand = {
    ...cmsFunctionServerCommand,
    describe,
    handler,
    builder,
};
export default deprecatedFunctionServerCommand;
